// @chainlink/data-streams-sdk/dist/tests/integration/context-cancellation.test.js
"use strict";
/**
 * Integration Tests for Context Cancellation
 *
 * Tests proper handling of cancelled operations for both REST and streaming APIs.
 * Validates graceful shutdown and resource cleanup when operations are aborted.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const src_1 = require("../../src");
const mockWebSocketServer_1 = require("../utils/mockWebSocketServer");
(0, globals_1.describe)("Context Cancellation Integration Tests", () => {
    let mockServer;
    let client;
    let stream = null;
    (0, globals_1.beforeEach)(async () => {
        mockServer = new mockWebSocketServer_1.MockWebSocketServer();
        await mockServer.waitForReady();
        const serverAddress = mockServer.getAddress();
        client = (0, src_1.createClient)({
            apiKey: "test-api-key",
            userSecret: "test-user-secret",
            endpoint: `http://${serverAddress}`,
            wsEndpoint: `ws://${serverAddress}`,
            logging: {
                logger: {
                    info: () => { },
                    error: () => { },
                    debug: () => { },
                    warn: () => { },
                },
            },
        });
    });
    (0, globals_1.afterEach)(async () => {
        if (stream) {
            await stream.close();
            stream = null;
        }
        await mockServer.close();
    });
    (0, globals_1.describe)("Stream connection cancellation", () => {
        (0, globals_1.it)("should handle cancellation during connection establishment", async () => {
            // Create stream but don't await connection
            stream = client.createStream(["0x0003" + "1".repeat(60)]);
            // Start connection process
            const connectionPromise = stream.connect();
            // Cancel immediately - this simulates context cancellation during connection
            await stream.close();
            // Connection should be aborted cleanly
            await (0, globals_1.expect)(connectionPromise).rejects.toThrow();
            // Stream should be in closed state
            const metrics = stream.getMetrics();
            (0, globals_1.expect)(metrics.activeConnections).toBe(0);
        }, 10000);
        (0, globals_1.it)("should handle cancellation after partial connection establishment", async () => {
            stream = client.createStream(["0x0003" + "1".repeat(60)]);
            // Let connection start
            await stream.connect();
            await new Promise(resolve => setTimeout(resolve, 100));
            // Verify connection was established
            const initialMetrics = stream.getMetrics();
            (0, globals_1.expect)(initialMetrics.activeConnections).toBeGreaterThan(0);
            // Cancel the connection
            await stream.close();
            // Verify clean shutdown
            const finalMetrics = stream.getMetrics();
            (0, globals_1.expect)(finalMetrics.activeConnections).toBe(0);
        }, 10000);
    });
    (0, globals_1.describe)("Stream operation cancellation", () => {
        (0, globals_1.it)("should stop receiving reports after stream closure", async () => {
            stream = client.createStream(["0x0003" + "1".repeat(60)]);
            let reportCount = 0;
            stream.on("report", () => {
                reportCount++;
            });
            await stream.connect();
            await new Promise(resolve => setTimeout(resolve, 100));
            // Send some test reports
            const mockReport = {
                report: {
                    feedID: "0x0003" + "1".repeat(60),
                    fullReport: "0x0002" + "01020304" + "42".repeat(100),
                    validFromTimestamp: Date.now(),
                    observationsTimestamp: Date.now(),
                },
            };
            await mockServer.broadcast(Buffer.from(JSON.stringify(mockReport)));
            await new Promise(resolve => setTimeout(resolve, 200));
            const reportsBeforeClose = reportCount;
            (0, globals_1.expect)(reportsBeforeClose).toBeGreaterThan(0);
            // Close the stream
            await stream.close();
            // Send more reports - these should not be received
            await mockServer.broadcast(Buffer.from(JSON.stringify(mockReport)));
            await new Promise(resolve => setTimeout(resolve, 200));
            // Report count should not increase after closure
            (0, globals_1.expect)(reportCount).toBe(reportsBeforeClose);
        }, 15000);
        (0, globals_1.it)("should handle rapid connect/disconnect cycles gracefully", async () => {
            stream = client.createStream(["0x0003" + "1".repeat(60)]);
            // Rapid connect/disconnect cycle
            for (let i = 0; i < 3; i++) {
                await stream.connect();
                await new Promise(resolve => setTimeout(resolve, 50));
                await stream.close();
                await new Promise(resolve => setTimeout(resolve, 50));
            }
            // Final state should be disconnected
            const metrics = stream.getMetrics();
            (0, globals_1.expect)(metrics.activeConnections).toBe(0);
        }, 15000);
    });
    (0, globals_1.describe)("Error handling during cancellation", () => {
        (0, globals_1.it)("should handle cancellation when server is unavailable", async () => {
            // Close the mock server to simulate unavailable server
            await mockServer.close();
            stream = client.createStream(["0x0003" + "1".repeat(60)]);
            // Try to connect to unavailable server
            const connectionPromise = stream.connect();
            // Cancel during failed connection attempt
            setTimeout(() => stream?.close(), 100);
            // Should handle cancellation gracefully even with connection failure
            await (0, globals_1.expect)(connectionPromise).rejects.toThrow();
            const metrics = stream.getMetrics();
            (0, globals_1.expect)(metrics.activeConnections).toBe(0);
        }, 10000);
        (0, globals_1.it)("should not leak resources after cancellation", async () => {
            const streams = [];
            // Create multiple streams and cancel them
            for (let i = 0; i < 5; i++) {
                const testStream = client.createStream(["0x0003" + i.toString().repeat(60)]);
                streams.push(testStream);
                // Start connection
                const connectPromise = testStream.connect();
                // Cancel after short delay
                setTimeout(() => testStream.close(), 50 + i * 10);
                try {
                    await connectPromise;
                }
                catch {
                    // Expected to fail due to cancellation
                }
            }
            // Wait for all operations to complete
            await new Promise(resolve => setTimeout(resolve, 1000));
            // All streams should be properly closed
            for (const testStream of streams) {
                const metrics = testStream.getMetrics();
                (0, globals_1.expect)(metrics.activeConnections).toBe(0);
            }
        }, 15000);
    });
    (0, globals_1.describe)("Event cleanup on cancellation", () => {
        (0, globals_1.it)("should stop emitting events after cancellation", async () => {
            stream = client.createStream(["0x0003" + "1".repeat(60)]);
            let eventCount = 0;
            const events = [];
            stream.on("report", () => {
                eventCount++;
                events.push("report");
            });
            stream.on("error", () => {
                eventCount++;
                events.push("error");
            });
            stream.on("disconnected", () => {
                eventCount++;
                events.push("disconnected");
            });
            await stream.connect();
            await new Promise(resolve => setTimeout(resolve, 100));
            const eventsBeforeClose = eventCount;
            // Close the stream
            await stream.close();
            // Trigger potential events by closing server
            await mockServer.close();
            await new Promise(resolve => setTimeout(resolve, 500));
            // Event count should not increase significantly after closure
            // (disconnected event might be emitted once)
            (0, globals_1.expect)(eventCount - eventsBeforeClose).toBeLessThanOrEqual(1);
        }, 10000);
    });
});
