// @chainlink/data-streams-sdk/dist/src/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.X_CLL_ORIGIN_HEADER = exports.X_CLL_AVAILABLE_ORIGINS_HEADER = exports.DEFAULT_RETRY_ATTEMPTS = exports.DEFAULT_RETRY_DELAY = exports.DEFAULT_TIMEOUT = exports.VALIDATION_REGEX = exports.WS_CONSTANTS = void 0;
/**
 * WebSocket connection constants
 */
exports.WS_CONSTANTS = {
    /** Maximum time to wait for connection in milliseconds */
    CONNECT_TIMEOUT: 5000,
    /** Ping interval in milliseconds */
    PING_INTERVAL: 30000,
    /** Time to wait for pong response in milliseconds */
    PONG_TIMEOUT: 5000,
    /** Maximum reconnection attempts */
    MAX_RECONNECTS: 5,
    /** Base delay between reconnection attempts in milliseconds */
    RECONNECT_DELAY: 1000,
    /** Maximum delay between reconnection attempts in milliseconds */
    MAX_RECONNECT_INTERVAL: 10000,
};
/**
 * Regular expressions for validation
 */
exports.VALIDATION_REGEX = {
    /** Matches valid feed IDs (0x followed by 64 hex characters) */
    FEED_ID: /^0x[0-9a-fA-F]{64}$/,
    /** Matches valid schema versions (0x0002-0x0009, 0x000a, 0x000d) */
    SCHEMA_VERSION: /^0x000([2-9]|a|b|c|d)$/,
};
// Request timeout constants
exports.DEFAULT_TIMEOUT = 30000; // 30 seconds
exports.DEFAULT_RETRY_DELAY = 1000; // 1 second
exports.DEFAULT_RETRY_ATTEMPTS = 1;
// HA Mode Constants
exports.X_CLL_AVAILABLE_ORIGINS_HEADER = "X-Cll-Available-Origins";
exports.X_CLL_ORIGIN_HEADER = "X-Cll-Origin";
