// @chainlink/data-streams-sdk/dist/src/utils/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SDKLogger = void 0;
const logger_1 = require("../types/logger");
/**
 * Simple and efficient internal logger for the Data Streams SDK.
 *
 * Provides configurable logging
 *
 * @example Basic usage
 * ```typescript
 * const logger = new SDKLogger({
 *   logger: { info: console.log, error: console.error },
 *   logLevel: LogLevel.INFO
 * });
 * logger.info('Client initialized');
 * ```
 *
 * @example Debugging WebSocket connections
 * ```typescript
 * const logger = new SDKLogger({
 *   logger: myLogger,
 *   enableConnectionDebug: true
 * });
 * logger.connectionDebug('Ping received from origin');
 * ```
 */
class SDKLogger {
    constructor(config = {}) {
        this.config = config;
    }
    /** General debug logs */
    debug(message, ...args) {
        this.log(logger_1.LogLevel.DEBUG, message, ...args);
    }
    /** Information logs */
    info(message, ...args) {
        this.log(logger_1.LogLevel.INFO, message, ...args);
    }
    /** Warning logs */
    warn(message, ...args) {
        this.log(logger_1.LogLevel.WARN, message, ...args);
    }
    /** Error logs */
    error(message, ...args) {
        this.log(logger_1.LogLevel.ERROR, message, ...args);
    }
    /** Specialized debug logs for WebSocket connections */
    connectionDebug(message, ...args) {
        if (this.config.enableConnectionDebug) {
            this.debug(`[Connection] ${message}`, ...args);
        }
    }
    /** Internal logging method with level verification */
    log(level, message, ...args) {
        // Zero overhead if no logger configured
        if (!this.config.logger) {
            return;
        }
        // Check minimum level
        const minLevel = this.config.logLevel ?? logger_1.LogLevel.INFO;
        if (level < minLevel) {
            return;
        }
        // Route to appropriate function
        const timestamp = new Date().toISOString();
        const formattedMessage = `[${timestamp}] [DataStreams] ${message}`;
        try {
            switch (level) {
                case logger_1.LogLevel.DEBUG:
                    this.config.logger.debug?.(formattedMessage, ...args);
                    break;
                case logger_1.LogLevel.INFO:
                    this.config.logger.info?.(formattedMessage, ...args);
                    break;
                case logger_1.LogLevel.WARN:
                    this.config.logger.warn?.(formattedMessage, ...args);
                    break;
                case logger_1.LogLevel.ERROR:
                    this.config.logger.error?.(formattedMessage, ...args);
                    break;
            }
        }
        catch {
            // Silent if external logger fails
            // Do not crash SDK due to logging issues
        }
    }
}
exports.SDKLogger = SDKLogger;
