// @chainlink/data-streams-sdk/dist/src/index.js
"use strict";
/**
 * Chainlink Data Streams TypeScript SDK
 *
 * A comprehensive SDK for accessing Chainlink Data Streams with full developer control.
 * Features event-driven architecture, High Availability mode, automatic failover,
 * and monitoring capabilities.
 *
 * @example Basic Usage
 * ```typescript
 * import { createClient } from '@chainlink/data-streams-sdk';
 *
 * const client = createClient({
 *   apiKey: 'your_api_key',
 *   userSecret: 'your_user_secret',
 *   endpoint: 'https://api.testnet-dataengine.chain.link',
 *   wsEndpoint: 'wss://ws.testnet-dataengine.chain.link'
 * });
 *
 * // Event-driven streaming with full developer control
 * const stream = client.createStream(['0x00037da06d56d083670...']);
 * stream.on('report', (report) => {
 *   console.log(`Price: ${report.price}, Feed: ${report.feedID}`);
 * });
 * stream.on('error', (error) => console.error('Error:', error));
 * await stream.connect();
 * ```
 *
 * @example High Availability Mode
 * ```typescript
 * const client = createClient({
 *   // ... auth config
 *   wsEndpoint: 'wss://ws1.example.com,wss://ws2.example.com',
 *   haMode: true,
 * });
 *
 * const stream = client.createStream(feedIds);
 * stream.on('report', (report) => processReport(report));
 * await stream.connect();
 * ```
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_RETRY_DELAY = exports.DEFAULT_RETRY_ATTEMPTS = exports.DEFAULT_TIMEOUT = exports.DEFAULT_CONFIG = exports.generateAuthHeaders = exports.getCurrentTimestamp = exports.formatReport = exports.getReportVersion = exports.ConnectionStatus = exports.LogLevel = exports.Stream = exports.decodeReport = exports.createClient = void 0;
// Core functionality
var client_1 = require("./client");
Object.defineProperty(exports, "createClient", { enumerable: true, get: function () { return client_1.createClient; } });
var decoder_1 = require("./decoder");
Object.defineProperty(exports, "decodeReport", { enumerable: true, get: function () { return decoder_1.decodeReport; } });
var stream_1 = require("./stream");
Object.defineProperty(exports, "Stream", { enumerable: true, get: function () { return stream_1.Stream; } });
var logger_1 = require("./types/logger");
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return logger_1.LogLevel; } });
var metrics_1 = require("./types/metrics");
Object.defineProperty(exports, "ConnectionStatus", { enumerable: true, get: function () { return metrics_1.ConnectionStatus; } });
__exportStar(require("./types/errors"), exports);
// Utility Functions
var utils_1 = require("./utils");
// Report utilities
Object.defineProperty(exports, "getReportVersion", { enumerable: true, get: function () { return utils_1.getReportVersion; } });
Object.defineProperty(exports, "formatReport", { enumerable: true, get: function () { return utils_1.formatReport; } });
// Time utilities
Object.defineProperty(exports, "getCurrentTimestamp", { enumerable: true, get: function () { return utils_1.getCurrentTimestamp; } });
// Authentication Headers
Object.defineProperty(exports, "generateAuthHeaders", { enumerable: true, get: function () { return utils_1.generateAuthHeaders; } });
// Constants
var defaultConfig_1 = require("./defaultConfig");
Object.defineProperty(exports, "DEFAULT_CONFIG", { enumerable: true, get: function () { return defaultConfig_1.DEFAULT_CONFIG; } });
var constants_1 = require("./utils/constants");
Object.defineProperty(exports, "DEFAULT_TIMEOUT", { enumerable: true, get: function () { return constants_1.DEFAULT_TIMEOUT; } });
Object.defineProperty(exports, "DEFAULT_RETRY_ATTEMPTS", { enumerable: true, get: function () { return constants_1.DEFAULT_RETRY_ATTEMPTS; } });
Object.defineProperty(exports, "DEFAULT_RETRY_DELAY", { enumerable: true, get: function () { return constants_1.DEFAULT_RETRY_DELAY; } });
