// @chainlink/data-streams-sdk/dist/tests/features/error-handling.test.js
"use strict";
/**
 * Comprehensive Error Handling Tests
 *
 * These tests validate the error handling system by:
 * - Testing all custom error types and their properties
 * - Testing error inheritance and instanceof checks
 * - Testing error message formatting and clarity
 * - Testing error context preservation across operations
 * - Testing error recovery scenarios and graceful degradation
 * - Testing error propagation through async operations
 *
 * Goals:
 * - Ensure robust error handling that provides clear debugging information
 * - Test all error scenarios comprehensively
 * - Validate error recovery and graceful degradation
 * - Provide excellent developer experience with meaningful error messages
 * - Build the best possible TypeScript error handling system
 */
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const errors_1 = require("../../src/types/errors");
(0, globals_1.describe)("Error Handling Tests", () => {
    (0, globals_1.describe)("custom error types", () => {
        (0, globals_1.describe)("DataStreamsError (base class)", () => {
            (0, globals_1.it)("should create error with message", () => {
                const error = new errors_1.DataStreamsError("Test error message");
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error.name).toBe("DataStreamsError");
                (0, globals_1.expect)(error.message).toBe("Test error message");
                (0, globals_1.expect)(error.stack).toBeDefined();
            });
            (0, globals_1.it)("should preserve stack trace", () => {
                const error = new errors_1.DataStreamsError("Test error");
                (0, globals_1.expect)(error.stack).toBeDefined();
                (0, globals_1.expect)(error.stack).toContain("DataStreamsError");
                (0, globals_1.expect)(error.stack).toContain("Test error");
            });
        });
        (0, globals_1.describe)("ValidationError", () => {
            (0, globals_1.it)("should create validation error", () => {
                const error = new errors_1.ValidationError("Invalid feed ID format");
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.ValidationError);
                (0, globals_1.expect)(error.name).toBe("ValidationError");
                (0, globals_1.expect)(error.message).toBe("Invalid feed ID format");
            });
            (0, globals_1.it)("should handle feed ID validation errors", () => {
                const feedId = "invalid-feed-id";
                const error = new errors_1.ValidationError(`Feed ID must be a valid hex string starting with 0x: ${feedId}`);
                (0, globals_1.expect)(error.message).toContain("Feed ID must be a valid hex string");
                (0, globals_1.expect)(error.message).toContain(feedId);
            });
            (0, globals_1.it)("should handle timestamp validation errors", () => {
                const timestamp = -1;
                const error = new errors_1.ValidationError(`Timestamp cannot be negative: ${timestamp}`);
                (0, globals_1.expect)(error.message).toContain("Timestamp cannot be negative");
                (0, globals_1.expect)(error.message).toContain(String(timestamp));
            });
        });
        (0, globals_1.describe)("AuthenticationError", () => {
            (0, globals_1.it)("should create authentication error", () => {
                const error = new errors_1.AuthenticationError("Invalid API key");
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.AuthenticationError);
                (0, globals_1.expect)(error.name).toBe("AuthenticationError");
                (0, globals_1.expect)(error.message).toBe("Invalid API key");
            });
            (0, globals_1.it)("should handle signature validation errors", () => {
                const error = new errors_1.AuthenticationError("HMAC signature validation failed");
                (0, globals_1.expect)(error.message).toContain("HMAC signature validation failed");
            });
            (0, globals_1.it)("should handle timestamp skew errors", () => {
                const error = new errors_1.AuthenticationError("Request timestamp too old or too far in the future");
                (0, globals_1.expect)(error.message).toContain("Request timestamp");
            });
        });
        (0, globals_1.describe)("APIError", () => {
            (0, globals_1.it)("should create API error", () => {
                const error = new errors_1.APIError("Connection timeout");
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.APIError);
                (0, globals_1.expect)(error.name).toBe("APIError");
                (0, globals_1.expect)(error.message).toBe("Connection timeout");
            });
            (0, globals_1.it)("should create API error with status code", () => {
                const error = new errors_1.APIError("HTTP 500: Internal Server Error", 500);
                (0, globals_1.expect)(error.message).toBe("HTTP 500: Internal Server Error");
                (0, globals_1.expect)(error.statusCode).toBe(500);
            });
            (0, globals_1.it)("should handle missing status code", () => {
                const error = new errors_1.APIError("Network unreachable");
                (0, globals_1.expect)(error.message).toBe("Network unreachable");
                (0, globals_1.expect)(error.statusCode).toBeUndefined();
            });
        });
        (0, globals_1.describe)("ReportDecodingError", () => {
            (0, globals_1.it)("should create report decoding error", () => {
                const error = new errors_1.ReportDecodingError("Invalid report format");
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.ReportDecodingError);
                (0, globals_1.expect)(error.name).toBe("ReportDecodingError");
                (0, globals_1.expect)(error.message).toBe("Invalid report format");
            });
            (0, globals_1.it)("should handle ABI decoding errors", () => {
                const error = new errors_1.ReportDecodingError("Failed to decode V3 report: insufficient data");
                (0, globals_1.expect)(error.message).toContain("Failed to decode V3 report");
                (0, globals_1.expect)(error.message).toContain("insufficient data");
            });
            (0, globals_1.it)("should handle version validation errors", () => {
                const version = "0x0099";
                const error = new errors_1.ReportDecodingError(`Unknown report version: ${version}`);
                (0, globals_1.expect)(error.message).toContain("Unknown report version");
                (0, globals_1.expect)(error.message).toContain(version);
            });
        });
        (0, globals_1.describe)("WebSocketError", () => {
            (0, globals_1.it)("should create WebSocket error", () => {
                const error = new errors_1.WebSocketError("WebSocket connection failed");
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.WebSocketError);
                (0, globals_1.expect)(error.name).toBe("WebSocketError");
                (0, globals_1.expect)(error.message).toBe("WebSocket connection failed");
            });
            (0, globals_1.it)("should handle connection errors", () => {
                const url = "wss://example.com";
                const error = new errors_1.WebSocketError(`Failed to connect to ${url}`);
                (0, globals_1.expect)(error.message).toContain("Failed to connect to");
                (0, globals_1.expect)(error.message).toContain(url);
            });
            (0, globals_1.it)("should handle message parsing errors", () => {
                const error = new errors_1.WebSocketError("Invalid message format received from stream");
                (0, globals_1.expect)(error.message).toContain("Invalid message format");
            });
        });
        (0, globals_1.describe)("OriginDiscoveryError", () => {
            (0, globals_1.it)("should create origin discovery error", () => {
                const error = new errors_1.OriginDiscoveryError("Failed to discover origins");
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.OriginDiscoveryError);
                (0, globals_1.expect)(error.name).toBe("OriginDiscoveryError");
                (0, globals_1.expect)(error.message).toBe("Failed to discover origins");
            });
            (0, globals_1.it)("should create origin discovery error with cause", () => {
                const cause = new Error("Network timeout");
                const error = new errors_1.OriginDiscoveryError("Failed to discover origins", cause);
                (0, globals_1.expect)(error.message).toBe("Failed to discover origins");
                (0, globals_1.expect)(error.cause).toBe(cause);
            });
            (0, globals_1.it)("should handle HEAD request failures", () => {
                const url = "https://api.example.com";
                const error = new errors_1.OriginDiscoveryError(`HEAD request failed for ${url}: 404 Not Found`);
                (0, globals_1.expect)(error.message).toContain("HEAD request failed");
                (0, globals_1.expect)(error.message).toContain(url);
                (0, globals_1.expect)(error.message).toContain("404 Not Found");
            });
            (0, globals_1.it)("should handle header parsing errors", () => {
                const error = new errors_1.OriginDiscoveryError("Invalid X-Cll-Available-Origins header format");
                (0, globals_1.expect)(error.message).toContain("X-Cll-Available-Origins header");
            });
        });
        (0, globals_1.describe)("MultiConnectionError", () => {
            (0, globals_1.it)("should create multi-connection error", () => {
                const error = new errors_1.MultiConnectionError("Failed to establish multiple connections");
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.MultiConnectionError);
                (0, globals_1.expect)(error.name).toBe("MultiConnectionError");
                (0, globals_1.expect)(error.message).toBe("Failed to establish multiple connections");
            });
            (0, globals_1.it)("should handle connection failure details", () => {
                const failedOrigins = ["wss://origin1.example.com", "wss://origin2.example.com"];
                const error = new errors_1.MultiConnectionError(`Failed to connect to origins: ${failedOrigins.join(", ")}`);
                (0, globals_1.expect)(error.message).toContain("Failed to connect to origins");
                (0, globals_1.expect)(error.message).toContain("origin1.example.com");
                (0, globals_1.expect)(error.message).toContain("origin2.example.com");
            });
        });
        (0, globals_1.describe)("PartialConnectionFailureError", () => {
            (0, globals_1.it)("should create partial connection failure error", () => {
                const error = new errors_1.PartialConnectionFailureError("Some connections failed", 2, 4);
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.PartialConnectionFailureError);
                (0, globals_1.expect)(error.name).toBe("PartialConnectionFailureError");
                (0, globals_1.expect)(error.message).toBe("Some connections failed");
                (0, globals_1.expect)(error.failedConnections).toBe(2);
                (0, globals_1.expect)(error.totalConnections).toBe(4);
            });
            (0, globals_1.it)("should handle partial failure details", () => {
                const activeCount = 2;
                const totalCount = 4;
                const error = new errors_1.PartialConnectionFailureError(`Partial connection failure: ${activeCount}/${totalCount} connections active`, totalCount - activeCount, totalCount);
                (0, globals_1.expect)(error.message).toContain("Partial connection failure");
                (0, globals_1.expect)(error.message).toContain("2/4 connections active");
                (0, globals_1.expect)(error.failedConnections).toBe(2);
                (0, globals_1.expect)(error.totalConnections).toBe(4);
            });
        });
        (0, globals_1.describe)("InsufficientConnectionsError", () => {
            (0, globals_1.it)("should create insufficient connections error", () => {
                const error = new errors_1.InsufficientConnectionsError("No active connections available", 0, 2);
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.InsufficientConnectionsError);
                (0, globals_1.expect)(error.name).toBe("InsufficientConnectionsError");
                (0, globals_1.expect)(error.message).toBe("No active connections available");
                (0, globals_1.expect)(error.availableConnections).toBe(0);
                (0, globals_1.expect)(error.requiredConnections).toBe(2);
            });
            (0, globals_1.it)("should handle minimum connection requirements", () => {
                const required = 2;
                const available = 0;
                const error = new errors_1.InsufficientConnectionsError(`Insufficient connections: need ${required}, have ${available}`, available, required);
                (0, globals_1.expect)(error.message).toContain("Insufficient connections");
                (0, globals_1.expect)(error.message).toContain("need 2, have 0");
                (0, globals_1.expect)(error.availableConnections).toBe(0);
                (0, globals_1.expect)(error.requiredConnections).toBe(2);
            });
        });
    });
    (0, globals_1.describe)("error inheritance and instanceof checks", () => {
        (0, globals_1.it)("should properly inherit from base Error class", () => {
            const errors = [
                new errors_1.DataStreamsError("test"),
                new errors_1.ValidationError("test"),
                new errors_1.AuthenticationError("test"),
                new errors_1.ReportDecodingError("test"),
                new errors_1.WebSocketError("test"),
                new errors_1.APIError("test"),
                new errors_1.OriginDiscoveryError("test"),
                new errors_1.MultiConnectionError("test"),
                new errors_1.PartialConnectionFailureError("test", 1, 2),
                new errors_1.InsufficientConnectionsError("test", 0, 1),
            ];
            errors.forEach(error => {
                (0, globals_1.expect)(error).toBeInstanceOf(Error);
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.DataStreamsError);
                (0, globals_1.expect)(error.name).toBeDefined();
                (0, globals_1.expect)(error.message).toBeDefined();
                (0, globals_1.expect)(error.stack).toBeDefined();
            });
        });
        (0, globals_1.it)("should support instanceof checks for specific error types", () => {
            const validationError = new errors_1.ValidationError("validation error");
            const authError = new errors_1.AuthenticationError("auth error");
            const apiError = new errors_1.APIError("api error");
            (0, globals_1.expect)(validationError instanceof errors_1.ValidationError).toBe(true);
            (0, globals_1.expect)(validationError instanceof errors_1.AuthenticationError).toBe(false);
            (0, globals_1.expect)(validationError instanceof errors_1.APIError).toBe(false);
            (0, globals_1.expect)(authError instanceof errors_1.AuthenticationError).toBe(true);
            (0, globals_1.expect)(authError instanceof errors_1.ValidationError).toBe(false);
            (0, globals_1.expect)(authError instanceof errors_1.APIError).toBe(false);
            (0, globals_1.expect)(apiError instanceof errors_1.APIError).toBe(true);
            (0, globals_1.expect)(apiError instanceof errors_1.ValidationError).toBe(false);
            (0, globals_1.expect)(apiError instanceof errors_1.AuthenticationError).toBe(false);
        });
        (0, globals_1.it)("should support polymorphic error handling", () => {
            const errors = [
                new errors_1.ValidationError("validation error"),
                new errors_1.AuthenticationError("auth error"),
                new errors_1.APIError("api error"),
            ];
            errors.forEach(error => {
                (0, globals_1.expect)(error instanceof errors_1.DataStreamsError).toBe(true);
                (0, globals_1.expect)(error instanceof Error).toBe(true);
                // Should be able to access base properties
                (0, globals_1.expect)(typeof error.name).toBe("string");
                (0, globals_1.expect)(typeof error.message).toBe("string");
                (0, globals_1.expect)(error.stack).toBeDefined();
            });
        });
    });
    (0, globals_1.describe)("error message formatting", () => {
        (0, globals_1.it)("should provide clear and descriptive error messages", () => {
            const testCases = [
                {
                    error: new errors_1.ValidationError("Missing required configuration field: apiKey"),
                    expectedPatterns: ["Missing required", "apiKey"],
                },
                {
                    error: new errors_1.ValidationError("Feed ID must be a 64-character hex string starting with 0x"),
                    expectedPatterns: ["Feed ID", "64-character hex string", "0x"],
                },
                {
                    error: new errors_1.AuthenticationError("HMAC signature mismatch: expected abc123, got def456"),
                    expectedPatterns: ["HMAC signature mismatch", "expected", "got"],
                },
                {
                    error: new errors_1.APIError("HTTP 429: Rate limit exceeded. Retry after 60 seconds", 429),
                    expectedPatterns: ["HTTP 429", "Rate limit exceeded", "Retry after"],
                },
                {
                    error: new errors_1.ReportDecodingError("Failed to decode V3 report: missing bid field"),
                    expectedPatterns: ["Failed to decode", "V3 report", "missing bid field"],
                },
            ];
            testCases.forEach(({ error, expectedPatterns }) => {
                expectedPatterns.forEach(pattern => {
                    (0, globals_1.expect)(error.message).toContain(pattern);
                });
            });
        });
        (0, globals_1.it)("should include relevant context in error messages", () => {
            const feedId = "0x0003" + "1".repeat(60);
            const timestamp = 1640995200;
            const url = "https://api.example.com";
            const errors = [
                new errors_1.ValidationError(`Invalid feed ID format: ${feedId}`),
                new errors_1.ValidationError(`Timestamp cannot be negative: ${timestamp}`),
                new errors_1.APIError(`Failed to connect to ${url}: Connection refused`),
                new errors_1.OriginDiscoveryError(`Origin discovery failed for ${url}: 404 Not Found`),
            ];
            (0, globals_1.expect)(errors[0].message).toContain(feedId);
            (0, globals_1.expect)(errors[1].message).toContain(String(timestamp));
            (0, globals_1.expect)(errors[2].message).toContain(url);
            (0, globals_1.expect)(errors[3].message).toContain(url);
        });
        (0, globals_1.it)("should format error messages consistently", () => {
            const errors = [
                new errors_1.ValidationError("Invalid endpoint URL"),
                new errors_1.ValidationError("Invalid feed ID format"),
                new errors_1.AuthenticationError("Invalid API key"),
                new errors_1.APIError("Connection timeout"),
                new errors_1.ReportDecodingError("Invalid report format"),
            ];
            errors.forEach(error => {
                // Should start with a capital letter
                (0, globals_1.expect)(error.message.charAt(0)).toMatch(/[A-Z]/);
                // Should not end with a period (for consistency)
                (0, globals_1.expect)(error.message).not.toMatch(/\.$/);
                // Should be non-empty
                (0, globals_1.expect)(error.message.length).toBeGreaterThan(0);
            });
        });
    });
    (0, globals_1.describe)("Documentation Test: error recovery scenarios", () => {
        // These tests demonstrate example error handling patterns
        (0, globals_1.it)("should demonstrate graceful degradation on partial failures", () => {
            const partialError = new errors_1.PartialConnectionFailureError("2 of 4 connections failed", 2, 4);
            // Simulate recovery logic
            function handlePartialFailure(error) {
                if (error instanceof errors_1.PartialConnectionFailureError) {
                    return {
                        canContinue: true,
                        message: "Continuing with reduced connection count",
                    };
                }
                return { canContinue: false, message: "Cannot recover" };
            }
            const result = handlePartialFailure(partialError);
            (0, globals_1.expect)(result.canContinue).toBe(true);
            (0, globals_1.expect)(result.message).toContain("reduced connection count");
        });
        (0, globals_1.it)("should demonstrate retry logic patterns for recoverable errors", () => {
            const retryableErrors = [
                new errors_1.APIError("HTTP 500: Internal Server Error", 500),
                new errors_1.APIError("HTTP 502: Bad Gateway", 502),
                new errors_1.APIError("HTTP 503: Service Unavailable", 503),
                new errors_1.APIError("Connection timeout"),
            ];
            const nonRetryableErrors = [
                new errors_1.AuthenticationError("Invalid API key"),
                new errors_1.ValidationError("Invalid feed ID"),
                new errors_1.ValidationError("Missing endpoint URL"),
            ];
            function isRetryable(error) {
                if (error instanceof errors_1.APIError) {
                    return (error.statusCode === 500 ||
                        error.statusCode === 502 ||
                        error.statusCode === 503 ||
                        error.message.includes("timeout"));
                }
                return false;
            }
            retryableErrors.forEach(error => {
                (0, globals_1.expect)(isRetryable(error)).toBe(true);
            });
            nonRetryableErrors.forEach(error => {
                (0, globals_1.expect)(isRetryable(error)).toBe(false);
            });
        });
        (0, globals_1.it)("should handle fallback mechanisms", () => {
            const primaryError = new errors_1.OriginDiscoveryError("Failed to discover origins");
            function handleOriginDiscoveryFailure(error) {
                if (error instanceof errors_1.OriginDiscoveryError) {
                    return {
                        fallbackUrl: "wss://fallback.example.com",
                        message: "Using fallback URL due to origin discovery failure",
                    };
                }
                throw error;
            }
            const result = handleOriginDiscoveryFailure(primaryError);
            (0, globals_1.expect)(result.fallbackUrl).toBe("wss://fallback.example.com");
            (0, globals_1.expect)(result.message).toContain("fallback URL");
        });
    });
    (0, globals_1.describe)("error propagation through async operations", () => {
        (0, globals_1.it)("should demonstrate error wrapping with cause preservation", async () => {
            async function simulateOriginDiscovery() {
                try {
                    // Simulates system errors that can occur during network operations
                    const systemError = new Error("Network timeout");
                    systemError.name = "AbortError"; // Simulates fetch timeout
                    throw systemError;
                }
                catch (error) {
                    // SDK: wrap system errors while preserving original context
                    throw new errors_1.OriginDiscoveryError("Failed to discover origins during HA setup", error);
                }
            }
            // Verify the wrapped error maintains both high-level context and system details
            await (0, globals_1.expect)(simulateOriginDiscovery()).rejects.toThrow(errors_1.OriginDiscoveryError);
            try {
                await simulateOriginDiscovery();
            }
            catch (error) {
                // Type guard for proper TS handling
                (0, globals_1.expect)(error).toBeInstanceOf(errors_1.OriginDiscoveryError);
                const originError = error;
                // Validate SDK level error info
                (0, globals_1.expect)(originError.message).toBe("Failed to discover origins during HA setup");
                // Verify system error is preserved
                (0, globals_1.expect)(originError.cause).toBeDefined();
                (0, globals_1.expect)(originError.cause).toBeInstanceOf(Error);
                (0, globals_1.expect)(originError.cause.name).toBe("AbortError");
                (0, globals_1.expect)(originError.cause.message).toBe("Network timeout");
            }
        });
        (0, globals_1.it)("should handle concurrent error scenarios", async () => {
            async function failingTask(id, delay) {
                await new Promise(resolve => setTimeout(resolve, delay));
                throw new errors_1.APIError(`Task ${id} failed`);
            }
            const tasks = [failingTask(1, 10), failingTask(2, 20), failingTask(3, 5)];
            // Test Promise.allSettled behavior with custom errors
            const results = await Promise.allSettled(tasks);
            (0, globals_1.expect)(results).toHaveLength(3);
            results.forEach((result, index) => {
                (0, globals_1.expect)(result.status).toBe("rejected");
                if (result.status === "rejected") {
                    (0, globals_1.expect)(result.reason).toBeInstanceOf(errors_1.APIError);
                    (0, globals_1.expect)(result.reason.message).toContain(`Task ${index + 1} failed`);
                }
            });
        });
    });
    (0, globals_1.describe)("error serialization and debugging", () => {
        (0, globals_1.it)("should provide useful toString representation", () => {
            const error = new errors_1.ValidationError("Invalid endpoint URL");
            const stringRep = error.toString();
            (0, globals_1.expect)(stringRep).toContain("ValidationError");
            (0, globals_1.expect)(stringRep).toContain("Invalid endpoint URL");
        });
        (0, globals_1.it)("should be JSON serializable for logging", () => {
            const error = new errors_1.APIError("Connection timeout", 408);
            const serialized = JSON.stringify({
                name: error.name,
                message: error.message,
                statusCode: error.statusCode,
                stack: error.stack,
            });
            const parsed = JSON.parse(serialized);
            (0, globals_1.expect)(parsed.name).toBe("APIError");
            (0, globals_1.expect)(parsed.message).toBe("Connection timeout");
            (0, globals_1.expect)(parsed.statusCode).toBe(408);
            (0, globals_1.expect)(parsed.stack).toBeDefined();
        });
        (0, globals_1.it)("should maintain error information across JSON serialization", () => {
            const originalError = new errors_1.ValidationError("Invalid feed ID format");
            // Simulate logging/serialization
            const errorInfo = {
                type: originalError.constructor.name,
                name: originalError.name,
                message: originalError.message,
                timestamp: new Date().toISOString(),
            };
            const serialized = JSON.stringify(errorInfo);
            const deserialized = JSON.parse(serialized);
            (0, globals_1.expect)(deserialized.type).toBe("ValidationError");
            (0, globals_1.expect)(deserialized.name).toBe("ValidationError");
            (0, globals_1.expect)(deserialized.message).toBe("Invalid feed ID format");
            (0, globals_1.expect)(deserialized.timestamp).toBeDefined();
        });
    });
    (0, globals_1.describe)("Documentation Test: error handling best practices", () => {
        // These tests demonstrate recommended error handling patterns
        (0, globals_1.it)("should demonstrate error filtering and categorization patterns", () => {
            const errors = [
                new errors_1.ValidationError("Missing API key"),
                new errors_1.APIError("Connection timeout"),
                new errors_1.ValidationError("Invalid feed ID"),
                new errors_1.AuthenticationError("Expired token"),
                new errors_1.ReportDecodingError("Malformed report"),
            ];
            const userErrors = errors.filter(error => error instanceof errors_1.ValidationError);
            const systemErrors = errors.filter(error => error instanceof errors_1.APIError || error instanceof errors_1.ReportDecodingError);
            const securityErrors = errors.filter(error => error instanceof errors_1.AuthenticationError);
            (0, globals_1.expect)(userErrors).toHaveLength(2);
            (0, globals_1.expect)(systemErrors).toHaveLength(2);
            (0, globals_1.expect)(securityErrors).toHaveLength(1);
        });
        (0, globals_1.it)("should demonstrate error severity classification patterns", () => {
            function getErrorSeverity(error) {
                if (error instanceof errors_1.ValidationError) {
                    return "medium";
                }
                if (error instanceof errors_1.AuthenticationError) {
                    return "high";
                }
                if (error instanceof errors_1.APIError || error instanceof errors_1.ReportDecodingError) {
                    return "low";
                }
                if (error instanceof errors_1.InsufficientConnectionsError) {
                    return "critical";
                }
                return "medium";
            }
            const testCases = [
                { error: new errors_1.ValidationError("test"), expectedSeverity: "medium" },
                { error: new errors_1.AuthenticationError("test"), expectedSeverity: "high" },
                { error: new errors_1.APIError("test"), expectedSeverity: "low" },
                { error: new errors_1.InsufficientConnectionsError("test", 0, 1), expectedSeverity: "critical" },
            ];
            testCases.forEach(({ error, expectedSeverity }) => {
                (0, globals_1.expect)(getErrorSeverity(error)).toBe(expectedSeverity);
            });
        });
        (0, globals_1.it)("should support error aggregation for batch operations", () => {
            const errors = [
                new errors_1.ValidationError("Feed ID 1 invalid"),
                new errors_1.ValidationError("Feed ID 2 invalid"),
                new errors_1.APIError("Connection failed"),
            ];
            class BatchError extends errors_1.DataStreamsError {
                constructor(errors) {
                    super(`Batch operation failed with ${errors.length} errors`);
                    this.errors = errors;
                    this.name = "BatchError";
                }
            }
            const batchError = new BatchError(errors);
            (0, globals_1.expect)(batchError.errors).toHaveLength(3);
            (0, globals_1.expect)(batchError.message).toContain("3 errors");
            (0, globals_1.expect)(batchError.errors[0]).toBeInstanceOf(errors_1.ValidationError);
            (0, globals_1.expect)(batchError.errors[2]).toBeInstanceOf(errors_1.APIError);
        });
    });
});
