// @chainlink/data-streams-sdk/dist/tests/features/ha-client-features.test.js
"use strict";
/**
 * Unit Tests for HA Client Features
 *
 * These tests validate HA-specific client functionality:
 * - createStream method with HA configuration
 * - validateHAConfiguration method
 * - Enhanced error messages for HA failures
 * - Event-driven architecture integration
 *
 * Requirements:
 * - Unit tests with mocked dependencies
 * - No actual network connections
 * - Focus on HA client configuration and integration
 */
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const implementation_1 = require("../../src/client/implementation");
const errors_1 = require("../../src/types/errors");
// Mock console methods to avoid noise during tests
const originalConsoleWarn = console.warn;
const originalConsoleError = console.error;
(0, globals_1.describe)("HA Client Features Tests", () => {
    let mockConfig;
    (0, globals_1.beforeEach)(() => {
        globals_1.jest.clearAllMocks();
        // Mock console methods
        console.warn = globals_1.jest.fn();
        console.error = globals_1.jest.fn();
        mockConfig = {
            apiKey: "test_key",
            userSecret: "test_secret",
            endpoint: "https://api.example.com",
            wsEndpoint: "wss://ws.example.com",
        };
    });
    afterEach(() => {
        // Restore console methods
        console.warn = originalConsoleWarn;
        console.error = originalConsoleError;
    });
    (0, globals_1.describe)("validateHAConfiguration", () => {
        (0, globals_1.it)("should validate HA mode with single origin without forced logging", () => {
            const haConfig = {
                ...mockConfig,
                wsEndpoint: "wss://single-origin.example.com",
                haMode: true,
            };
            // Should not throw and should NOT log anything (developers control logging)
            (0, globals_1.expect)(() => new implementation_1.DataStreamsClientImpl(haConfig)).not.toThrow();
            (0, globals_1.expect)(console.warn).not.toHaveBeenCalled();
        });
        (0, globals_1.it)("should validate HA mode with multiple origins without warning", () => {
            const haConfig = {
                ...mockConfig,
                wsEndpoint: "wss://origin1.example.com,wss://origin2.example.com",
                haMode: true,
            };
            (0, globals_1.expect)(() => new implementation_1.DataStreamsClientImpl(haConfig)).not.toThrow();
            (0, globals_1.expect)(console.warn).not.toHaveBeenCalled();
        });
        (0, globals_1.it)("should validate low HA connection timeout without forced logging", () => {
            const haConfig = {
                ...mockConfig,
                wsEndpoint: "wss://origin1.example.com,wss://origin2.example.com",
                haMode: true,
                haConnectionTimeout: 500, // Very low timeout
            };
            // Should not throw and should NOT log anything (developers control logging)
            (0, globals_1.expect)(() => new implementation_1.DataStreamsClientImpl(haConfig)).not.toThrow();
            (0, globals_1.expect)(console.warn).not.toHaveBeenCalled();
        });
        (0, globals_1.it)("should throw error for HA mode with no WebSocket endpoints", () => {
            const haConfig = {
                ...mockConfig,
                wsEndpoint: "",
                haMode: true,
            };
            // Base validation catches empty wsEndpoint first
            (0, globals_1.expect)(() => new implementation_1.DataStreamsClientImpl(haConfig)).toThrow("wsEndpoint cannot be empty");
        });
        (0, globals_1.it)("should throw error for invalid WebSocket URL format", () => {
            const haConfig = {
                ...mockConfig,
                wsEndpoint: "https://invalid-protocol.example.com,wss://valid.example.com",
                haMode: true,
            };
            // Base validation catches invalid protocol first
            (0, globals_1.expect)(() => new implementation_1.DataStreamsClientImpl(haConfig)).toThrow("wsEndpoint must use one of these protocols: ws:, wss:");
        });
        (0, globals_1.it)("should throw HA-specific error when URLs are parsed but empty", () => {
            // This tests the actual HA validation logic by bypassing base validation
            const haConfig = {
                ...mockConfig,
                wsEndpoint: "wss://valid.example.com", // Valid URL that passes base validation
                haMode: true,
            };
            // Mock parseOrigins to return empty array to trigger HA-specific validation
            const client = new implementation_1.DataStreamsClientImpl(haConfig);
            const parseOriginsSpy = globals_1.jest.spyOn(client, "parseOrigins");
            parseOriginsSpy.mockReturnValue([]);
            (0, globals_1.expect)(() => {
                // Call validateHAConfiguration directly to test HA-specific logic
                client.validateHAConfiguration({ ...haConfig, wsEndpoint: "wss://valid.example.com" });
            }).toThrow("HA mode enabled but no WebSocket endpoints provided");
        });
        (0, globals_1.it)("should validate comma-separated URLs correctly", () => {
            const haConfig = {
                ...mockConfig,
                wsEndpoint: "wss://origin1.example.com, wss://origin2.example.com , wss://origin3.example.com",
                haMode: true,
            };
            (0, globals_1.expect)(() => new implementation_1.DataStreamsClientImpl(haConfig)).not.toThrow();
        });
        (0, globals_1.it)("should handle mixed ws and wss protocols", () => {
            const haConfig = {
                ...mockConfig,
                wsEndpoint: "ws://localhost:8080,wss://production.example.com",
                haMode: true,
            };
            (0, globals_1.expect)(() => new implementation_1.DataStreamsClientImpl(haConfig)).not.toThrow();
        });
        (0, globals_1.it)("should warn appropriately when origin discovery is enabled with single origin", () => {
            const haConfig = {
                ...mockConfig,
                wsEndpoint: "wss://single-origin.example.com",
                haMode: true,
            };
            (0, globals_1.expect)(() => new implementation_1.DataStreamsClientImpl(haConfig)).not.toThrow();
            (0, globals_1.expect)(console.warn).not.toHaveBeenCalledWith(globals_1.expect.stringContaining("HA mode enabled but only one origin provided"));
        });
    });
    (0, globals_1.describe)("createStream method with HA features", () => {
        (0, globals_1.it)("should exist and have correct signature", () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            (0, globals_1.expect)(typeof client.createStream).toBe("function");
            (0, globals_1.expect)(client.createStream.length).toBe(2); // feedIds, options
        });
        (0, globals_1.it)("should create stream with proper configuration", () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            const feedIds = ["0x0003" + "1".repeat(60)];
            // Mock Stream constructor to capture configuration
            const StreamConstructorSpy = globals_1.jest.spyOn(require("../../src/stream"), "Stream");
            const mockStream = {
                on: globals_1.jest.fn().mockReturnThis(),
                connect: globals_1.jest.fn().mockImplementation(() => Promise.resolve()),
                close: globals_1.jest.fn().mockImplementation(() => Promise.resolve()),
                getMetrics: globals_1.jest.fn().mockReturnValue({}),
                getConnectionType: globals_1.jest.fn().mockReturnValue("single"),
                getOrigins: globals_1.jest.fn().mockReturnValue([]),
            };
            StreamConstructorSpy.mockImplementation((..._args) => mockStream);
            try {
                const stream = client.createStream(feedIds, {
                    maxReconnectAttempts: 10,
                    reconnectInterval: 5000,
                });
                // Verify Stream was created with correct parameters
                (0, globals_1.expect)(StreamConstructorSpy).toHaveBeenCalledWith(mockConfig, feedIds, {
                    maxReconnectAttempts: 10,
                    reconnectInterval: 5000,
                });
                (0, globals_1.expect)(stream).toBe(mockStream);
            }
            finally {
                StreamConstructorSpy.mockRestore();
            }
        });
        (0, globals_1.it)("should work with connection status callback in config", () => {
            const statusCallback = globals_1.jest.fn();
            const haConfig = {
                ...mockConfig,
                haMode: true,
                connectionStatusCallback: statusCallback,
            };
            const client = new implementation_1.DataStreamsClientImpl(haConfig);
            const feedIds = ["0x0003" + "1".repeat(60)];
            // Mock Stream constructor
            const StreamConstructorSpy = globals_1.jest.spyOn(require("../../src/stream"), "Stream");
            const mockStream = {
                on: globals_1.jest.fn().mockReturnThis(),
                connect: globals_1.jest.fn().mockImplementation(() => Promise.resolve()),
                close: globals_1.jest.fn().mockImplementation(() => Promise.resolve()),
                getMetrics: globals_1.jest.fn().mockReturnValue({}),
                getConnectionType: globals_1.jest.fn().mockReturnValue("multiple"),
                getOrigins: globals_1.jest.fn().mockReturnValue(["wss://ws.example.com"]),
            };
            StreamConstructorSpy.mockImplementation((..._args) => mockStream);
            try {
                client.createStream(feedIds);
                // Verify config includes the status callback
                const passedConfig = StreamConstructorSpy.mock.calls[0][0];
                (0, globals_1.expect)(passedConfig.connectionStatusCallback).toBe(statusCallback);
                (0, globals_1.expect)(passedConfig.haMode).toBe(true);
            }
            finally {
                StreamConstructorSpy.mockRestore();
            }
        });
    });
    (0, globals_1.describe)("event-driven error handling", () => {
        (0, globals_1.it)("should handle OriginDiscoveryError through stream events", async () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            // Create a mock stream that emits error events
            const mockStream = {
                on: globals_1.jest.fn(),
                connect: globals_1.jest.fn().mockImplementation(async () => {
                    throw new errors_1.OriginDiscoveryError("Discovery failed", new Error("Network timeout"));
                }),
                close: globals_1.jest.fn(),
                getConnectionType: globals_1.jest.fn().mockReturnValue("single"),
                getOrigins: globals_1.jest.fn().mockReturnValue([]),
                getMetrics: globals_1.jest.fn().mockReturnValue({}),
            };
            // Mock Stream constructor
            const StreamConstructorSpy = globals_1.jest.spyOn(require("../../src/stream"), "Stream");
            StreamConstructorSpy.mockImplementation((..._args) => mockStream);
            const feedIds = ["0x0003" + "1".repeat(60)];
            try {
                const stream = client.createStream(feedIds);
                // Test that errors can be handled via events
                const errorHandler = globals_1.jest.fn();
                stream.on("error", errorHandler);
                // When connecting fails, developers can handle the error via events
                await (0, globals_1.expect)(stream.connect()).rejects.toThrow(errors_1.OriginDiscoveryError);
                // Verify the stream object was created correctly
                (0, globals_1.expect)(stream).toBe(mockStream);
                (0, globals_1.expect)(mockStream.on).toHaveBeenCalledWith("error", errorHandler);
            }
            finally {
                StreamConstructorSpy.mockRestore();
            }
        });
        (0, globals_1.it)("should handle InsufficientConnectionsError through stream events", async () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            // Mock Stream that throws InsufficientConnectionsError
            const mockStream = {
                on: globals_1.jest.fn(),
                connect: globals_1.jest.fn().mockImplementation(async () => {
                    throw new errors_1.InsufficientConnectionsError("No connections", 0, 2);
                }),
                close: globals_1.jest.fn(),
                getConnectionType: globals_1.jest.fn().mockReturnValue("single"),
                getOrigins: globals_1.jest.fn().mockReturnValue([]),
                getMetrics: globals_1.jest.fn().mockReturnValue({}),
            };
            // Mock Stream constructor
            const StreamConstructorSpy = globals_1.jest.spyOn(require("../../src/stream"), "Stream");
            StreamConstructorSpy.mockImplementation((..._args) => mockStream);
            const feedIds = ["0x0003" + "1".repeat(60)];
            try {
                const stream = client.createStream(feedIds);
                // Test event-driven error handling
                const errorHandler = globals_1.jest.fn();
                stream.on("error", errorHandler);
                await (0, globals_1.expect)(stream.connect()).rejects.toThrow(errors_1.InsufficientConnectionsError);
                // Verify event handlers can be attached
                (0, globals_1.expect)(mockStream.on).toHaveBeenCalledWith("error", errorHandler);
            }
            finally {
                StreamConstructorSpy.mockRestore();
            }
        });
        (0, globals_1.it)("should support all event types for production monitoring", () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            // Mock complete stream with all event capabilities
            const mockStream = {
                on: globals_1.jest.fn().mockReturnThis(),
                connect: globals_1.jest.fn().mockImplementation(() => Promise.resolve()),
                close: globals_1.jest.fn().mockImplementation(() => Promise.resolve()),
                getMetrics: globals_1.jest.fn().mockReturnValue({
                    accepted: 100,
                    deduplicated: 10,
                    activeConnections: 2,
                    configuredConnections: 2,
                }),
                getConnectionType: globals_1.jest.fn().mockReturnValue("multiple"),
                getOrigins: globals_1.jest.fn().mockReturnValue(["wss://origin1.example.com", "wss://origin2.example.com"]),
            };
            const StreamConstructorSpy = globals_1.jest.spyOn(require("../../src/stream"), "Stream");
            StreamConstructorSpy.mockImplementation((..._args) => mockStream);
            try {
                const feedIds = ["0x0003" + "1".repeat(60)];
                const stream = client.createStream(feedIds);
                // Test all event types for production monitoring
                const reportHandler = globals_1.jest.fn();
                const errorHandler = globals_1.jest.fn();
                const disconnectedHandler = globals_1.jest.fn();
                const reconnectingHandler = globals_1.jest.fn();
                stream.on("report", reportHandler);
                stream.on("error", errorHandler);
                stream.on("disconnected", disconnectedHandler);
                stream.on("reconnecting", reconnectingHandler);
                // Verify all event handlers were registered
                (0, globals_1.expect)(mockStream.on).toHaveBeenCalledWith("report", reportHandler);
                (0, globals_1.expect)(mockStream.on).toHaveBeenCalledWith("error", errorHandler);
                (0, globals_1.expect)(mockStream.on).toHaveBeenCalledWith("disconnected", disconnectedHandler);
                (0, globals_1.expect)(mockStream.on).toHaveBeenCalledWith("reconnecting", reconnectingHandler);
                // Verify stream statistics are accessible
                const stats = stream.getMetrics();
                (0, globals_1.expect)(stats.accepted).toBe(100);
                (0, globals_1.expect)(stats.activeConnections).toBe(2);
            }
            finally {
                StreamConstructorSpy.mockRestore();
            }
        });
    });
    (0, globals_1.describe)("parseOrigins method", () => {
        (0, globals_1.it)("should correctly parse comma-separated URLs", () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            // Access private method for testing
            const parseOrigins = client.parseOrigins;
            const result = parseOrigins("wss://origin1.example.com,wss://origin2.example.com,wss://origin3.example.com");
            (0, globals_1.expect)(result).toEqual(["wss://origin1.example.com", "wss://origin2.example.com", "wss://origin3.example.com"]);
        });
        (0, globals_1.it)("should handle URLs with spaces", () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            const parseOrigins = client.parseOrigins;
            const result = parseOrigins(" wss://origin1.example.com , wss://origin2.example.com , wss://origin3.example.com ");
            (0, globals_1.expect)(result).toEqual(["wss://origin1.example.com", "wss://origin2.example.com", "wss://origin3.example.com"]);
        });
        (0, globals_1.it)("should filter out empty URLs", () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            const parseOrigins = client.parseOrigins;
            const result = parseOrigins("wss://origin1.example.com,,wss://origin3.example.com,");
            (0, globals_1.expect)(result).toEqual(["wss://origin1.example.com", "wss://origin3.example.com"]);
        });
        (0, globals_1.it)("should handle single URL", () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            const parseOrigins = client.parseOrigins;
            const result = parseOrigins("wss://single-origin.example.com");
            (0, globals_1.expect)(result).toEqual(["wss://single-origin.example.com"]);
        });
        (0, globals_1.it)("should return empty array for empty string", () => {
            const client = new implementation_1.DataStreamsClientImpl(mockConfig);
            const parseOrigins = client.parseOrigins;
            const result = parseOrigins("");
            (0, globals_1.expect)(result).toEqual([]);
        });
    });
});
