// @chainlink/data-streams-sdk/dist/src/types/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InsufficientConnectionsError = exports.PartialConnectionFailureError = exports.MultiConnectionError = exports.OriginDiscoveryError = exports.APIError = exports.WebSocketError = exports.ReportDecodingError = exports.AuthenticationError = exports.ValidationError = exports.DataStreamsError = void 0;
/**
 * Base error class for all Data Streams SDK errors.
 *
 * All SDK-specific errors inherit from this class, enabling easy error type detection
 * and unified error handling across the entire SDK.
 */
class DataStreamsError extends Error {
    constructor(message) {
        super(message);
        this.name = "DataStreamsError";
        Object.setPrototypeOf(this, DataStreamsError.prototype);
    }
}
exports.DataStreamsError = DataStreamsError;
/**
 * Thrown when input validation fails.
 *
 * Common scenarios include invalid feed IDs, empty required parameters,
 * invalid configuration values, or malformed timestamp values.
 */
class ValidationError extends DataStreamsError {
    constructor(message) {
        super(message);
        this.name = "ValidationError";
        Object.setPrototypeOf(this, ValidationError.prototype);
    }
}
exports.ValidationError = ValidationError;
/**
 * Thrown when authentication with Data Streams API fails.
 *
 * Common scenarios include invalid credentials, malformed HMAC signature,
 * expired tokens, or missing authentication headers.
 */
class AuthenticationError extends DataStreamsError {
    constructor(message) {
        super(message);
        this.name = "AuthenticationError";
        Object.setPrototypeOf(this, AuthenticationError.prototype);
    }
}
exports.AuthenticationError = AuthenticationError;
/**
 * Thrown when report data decoding fails.
 *
 * Common scenarios include corrupted report data, unsupported report format version,
 * or invalid report structure.
 */
class ReportDecodingError extends DataStreamsError {
    constructor(message) {
        super(message);
        this.name = "ReportDecodingError";
        Object.setPrototypeOf(this, ReportDecodingError.prototype);
    }
}
exports.ReportDecodingError = ReportDecodingError;
/**
 * Thrown when WebSocket connection establishment or management fails.
 *
 * Common scenarios include network connectivity issues, server unavailability,
 * connection timeouts, or protocol-level errors.
 *
 * Note: In HA mode, individual WebSocket failures may not interrupt the stream
 * if other connections remain active.
 */
class WebSocketError extends DataStreamsError {
    constructor(message) {
        super(message);
        this.name = "WebSocketError";
        Object.setPrototypeOf(this, WebSocketError.prototype);
    }
}
exports.WebSocketError = WebSocketError;
/**
 * Thrown when REST API requests fail.
 *
 * Common scenarios include HTTP 4xx/5xx errors, network timeouts,
 * rate limiting, or malformed requests.
 *
 * @param statusCode - HTTP status code from the failed request (if available)
 */
class APIError extends DataStreamsError {
    constructor(message, statusCode) {
        super(message);
        this.statusCode = statusCode;
        this.name = "APIError";
        Object.setPrototypeOf(this, APIError.prototype);
    }
}
exports.APIError = APIError;
/**
 * Thrown when automatic origin discovery fails in High Availability mode.
 *
 * This occurs when the client cannot discover additional WebSocket endpoints
 * via HEAD request to the origin server. The stream may fall back to static
 * configuration or single-connection mode.
 *
 * @param cause - The underlying error that caused discovery to fail (if available)
 */
class OriginDiscoveryError extends DataStreamsError {
    constructor(message, cause) {
        super(message);
        this.cause = cause;
        this.name = "OriginDiscoveryError";
        Object.setPrototypeOf(this, OriginDiscoveryError.prototype);
    }
}
exports.OriginDiscoveryError = OriginDiscoveryError;
/**
 * Thrown when all connections fail in High Availability mode.
 *
 * This is a critical error indicating complete failure of the HA connection system.
 * No WebSocket connections could be established to any discovered or configured origins.
 */
class MultiConnectionError extends DataStreamsError {
    constructor(message) {
        super(message);
        this.name = "MultiConnectionError";
        Object.setPrototypeOf(this, MultiConnectionError.prototype);
    }
}
exports.MultiConnectionError = MultiConnectionError;
/**
 * Thrown when some (but not all) connections fail in High Availability mode.
 *
 * This indicates degraded HA performance. The stream continues to operate with
 * successful connections, but redundancy is reduced.
 *
 * @param failedConnections - Number of connections that failed to establish
 * @param totalConnections - Total number of connections attempted
 */
class PartialConnectionFailureError extends DataStreamsError {
    constructor(message, failedConnections, totalConnections) {
        super(message);
        this.failedConnections = failedConnections;
        this.totalConnections = totalConnections;
        this.name = "PartialConnectionFailureError";
        Object.setPrototypeOf(this, PartialConnectionFailureError.prototype);
    }
}
exports.PartialConnectionFailureError = PartialConnectionFailureError;
/**
 * Thrown when insufficient connections are available for optimal High Availability mode.
 *
 * This warning-level error indicates that while some connections succeeded, the number
 * is below the threshold for robust HA operation. The stream continues with reduced
 * fault tolerance.
 *
 * @param availableConnections - Number of successful connections established
 * @param requiredConnections - Minimum number of connections desired for full HA
 */
class InsufficientConnectionsError extends DataStreamsError {
    constructor(message, availableConnections, requiredConnections) {
        super(message);
        this.availableConnections = availableConnections;
        this.requiredConnections = requiredConnections;
        this.name = "InsufficientConnectionsError";
        Object.setPrototypeOf(this, InsufficientConnectionsError.prototype);
    }
}
exports.InsufficientConnectionsError = InsufficientConnectionsError;
