// @chainlink/data-streams-sdk/dist/tests/unit/utils/validation/config-validation.test.js
"use strict";
/**
 * Unit Tests for Configuration Validation
 *
 * These tests validate the configuration functionality by:
 * - Testing valid config creation and validation
 * - Testing missing required fields (apiKey, userSecret, endpoint, wsEndpoint)
 * - Testing invalid URL formats (REST and WebSocket)
 * - Testing URL scheme validation
 * - Testing timeout and retry configuration validation
 * - Testing HA mode configuration validation
 * - Testing config edge cases and error messages
 * - Testing config normalization and defaults
 *
 * Goals:
 * - Ensure robust config validation that prevents invalid configurations
 * - Test all edge cases and error scenarios comprehensively
 * - Provide clear, helpful error messages for developers
 * - Build the best possible TypeScript configuration validation
 */
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const client_1 = require("../../../../src/client");
// Mock console methods to avoid noise during tests
jest.spyOn(console, "info").mockImplementation(() => { });
jest.spyOn(console, "warn").mockImplementation(() => { });
jest.spyOn(console, "error").mockImplementation(() => { });
jest.spyOn(console, "log").mockImplementation(() => { });
(0, globals_1.describe)("Configuration Validation Tests", () => {
    // Valid base configuration for testing
    const VALID_CONFIG = {
        apiKey: "test-api-key-12345",
        userSecret: "test-user-secret-67890",
        endpoint: "https://api.testnet-dataengine.chain.link",
        wsEndpoint: "wss://ws.testnet-dataengine.chain.link",
    };
    (0, globals_1.describe)("valid configuration creation", () => {
        (0, globals_1.it)("should accept minimal valid configuration", () => {
            (0, globals_1.expect)(() => (0, client_1.createClient)(VALID_CONFIG)).not.toThrow();
        });
        (0, globals_1.it)("should accept configuration with all optional fields", () => {
            const fullConfig = {
                ...VALID_CONFIG,
                retryAttempts: 5,
                retryDelay: 2000,
                timeout: 60000,
                haMode: true,
                haConnectionTimeout: 10000,
                logging: {
                    logger: console,
                },
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(fullConfig)).not.toThrow();
        });
        (0, globals_1.it)("should accept configuration with HA mode enabled", () => {
            const haConfig = {
                ...VALID_CONFIG,
                haMode: true,
                wsEndpoint: "wss://ws1.example.com,wss://ws2.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(haConfig)).not.toThrow();
        });
        (0, globals_1.it)("should accept configuration with single WebSocket URL", () => {
            const singleWsConfig = {
                ...VALID_CONFIG,
                wsEndpoint: "wss://ws.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(singleWsConfig)).not.toThrow();
        });
        (0, globals_1.it)("should accept configuration with multiple comma-separated WebSocket URLs", () => {
            const multiWsConfig = {
                ...VALID_CONFIG,
                wsEndpoint: "wss://ws1.example.com,wss://ws2.example.com,wss://ws3.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(multiWsConfig)).not.toThrow();
        });
    });
    (0, globals_1.describe)("required field validation", () => {
        (0, globals_1.it)("should reject configuration without apiKey", () => {
            const configWithoutApiKey = {
                ...VALID_CONFIG,
                apiKey: undefined,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithoutApiKey)).toThrow();
        });
        (0, globals_1.it)("should reject configuration with empty apiKey", () => {
            const configWithEmptyApiKey = {
                ...VALID_CONFIG,
                apiKey: "",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithEmptyApiKey)).toThrow();
        });
        (0, globals_1.it)("should reject configuration without userSecret", () => {
            const configWithoutUserSecret = {
                ...VALID_CONFIG,
                userSecret: undefined,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithoutUserSecret)).toThrow();
        });
        (0, globals_1.it)("should reject configuration with empty userSecret", () => {
            const configWithEmptyUserSecret = {
                ...VALID_CONFIG,
                userSecret: "",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithEmptyUserSecret)).toThrow();
        });
        (0, globals_1.it)("should reject configuration without endpoint", () => {
            const configWithoutEndpoint = {
                ...VALID_CONFIG,
                endpoint: undefined,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithoutEndpoint)).toThrow();
        });
        (0, globals_1.it)("should reject configuration with empty endpoint", () => {
            const configWithEmptyEndpoint = {
                ...VALID_CONFIG,
                endpoint: "",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithEmptyEndpoint)).toThrow();
        });
        (0, globals_1.it)("should reject configuration without wsEndpoint", () => {
            const configWithoutWsEndpoint = {
                ...VALID_CONFIG,
                wsEndpoint: undefined,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithoutWsEndpoint)).toThrow();
        });
        (0, globals_1.it)("should reject configuration with empty wsEndpoint", () => {
            const configWithEmptyWsEndpoint = {
                ...VALID_CONFIG,
                wsEndpoint: "",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithEmptyWsEndpoint)).toThrow();
        });
    });
    (0, globals_1.describe)("URL format validation", () => {
        (0, globals_1.it)("should reject invalid REST URL format", () => {
            const configWithInvalidRestUrl = {
                ...VALID_CONFIG,
                endpoint: "not-a-valid-url",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithInvalidRestUrl)).toThrow();
        });
        (0, globals_1.it)("should reject REST URL without protocol", () => {
            const configWithoutProtocol = {
                ...VALID_CONFIG,
                endpoint: "api.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithoutProtocol)).toThrow();
        });
        (0, globals_1.it)("should reject REST URL with invalid protocol", () => {
            const configWithInvalidProtocol = {
                ...VALID_CONFIG,
                endpoint: "ftp://api.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithInvalidProtocol)).toThrow();
        });
        (0, globals_1.it)("should accept HTTPS REST URLs", () => {
            const configWithHttps = {
                ...VALID_CONFIG,
                endpoint: "https://api.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithHttps)).not.toThrow();
        });
        (0, globals_1.it)("should accept HTTP REST URLs (for testing)", () => {
            const configWithHttp = {
                ...VALID_CONFIG,
                endpoint: "http://localhost:8080",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithHttp)).not.toThrow();
        });
        (0, globals_1.it)("should reject invalid WebSocket URL format", () => {
            const configWithInvalidWsUrl = {
                ...VALID_CONFIG,
                wsEndpoint: "not-a-valid-websocket-url",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithInvalidWsUrl)).toThrow();
        });
        (0, globals_1.it)("should reject WebSocket URL without ws/wss protocol", () => {
            const configWithoutWsProtocol = {
                ...VALID_CONFIG,
                wsEndpoint: "https://ws.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithoutWsProtocol)).toThrow();
        });
        (0, globals_1.it)("should accept WSS WebSocket URLs", () => {
            const configWithWss = {
                ...VALID_CONFIG,
                wsEndpoint: "wss://ws.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithWss)).not.toThrow();
        });
        (0, globals_1.it)("should accept WS WebSocket URLs (for testing)", () => {
            const configWithWs = {
                ...VALID_CONFIG,
                wsEndpoint: "ws://localhost:8080",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithWs)).not.toThrow();
        });
        (0, globals_1.it)("should reject malformed URLs in comma-separated list", () => {
            const configWithMalformedUrls = {
                ...VALID_CONFIG,
                wsEndpoint: "wss://ws1.example.com,invalid-url,wss://ws2.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithMalformedUrls)).toThrow();
        });
        (0, globals_1.it)("should reject mixed protocols in comma-separated list", () => {
            const configWithMixedProtocols = {
                ...VALID_CONFIG,
                wsEndpoint: "wss://ws1.example.com,https://ws2.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithMixedProtocols)).toThrow();
        });
    });
    (0, globals_1.describe)("timeout and retry configuration validation", () => {
        (0, globals_1.it)("should accept valid timeout values", () => {
            const configWithTimeout = {
                ...VALID_CONFIG,
                timeout: 30000,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithTimeout)).not.toThrow();
        });
        (0, globals_1.it)("should reject negative timeout", () => {
            const configWithNegativeTimeout = {
                ...VALID_CONFIG,
                timeout: -1000,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithNegativeTimeout)).toThrow();
        });
        (0, globals_1.it)("should reject zero timeout", () => {
            const configWithZeroTimeout = {
                ...VALID_CONFIG,
                timeout: 0,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithZeroTimeout)).toThrow();
        });
        (0, globals_1.it)("should accept valid retry attempts", () => {
            const configWithRetryAttempts = {
                ...VALID_CONFIG,
                retryAttempts: 5,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithRetryAttempts)).not.toThrow();
        });
        (0, globals_1.it)("should reject negative retry attempts", () => {
            const configWithNegativeRetry = {
                ...VALID_CONFIG,
                retryAttempts: -1,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithNegativeRetry)).toThrow();
        });
        (0, globals_1.it)("should accept zero retry attempts", () => {
            const configWithZeroRetry = {
                ...VALID_CONFIG,
                retryAttempts: 0,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithZeroRetry)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid retry delay", () => {
            const configWithRetryDelay = {
                ...VALID_CONFIG,
                retryDelay: 2000,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithRetryDelay)).not.toThrow();
        });
        (0, globals_1.it)("should reject negative retry delay", () => {
            const configWithNegativeDelay = {
                ...VALID_CONFIG,
                retryDelay: -500,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithNegativeDelay)).toThrow();
        });
        (0, globals_1.it)("should accept zero retry delay", () => {
            const configWithZeroDelay = {
                ...VALID_CONFIG,
                retryDelay: 0,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithZeroDelay)).not.toThrow();
        });
    });
    (0, globals_1.describe)("HA mode configuration validation", () => {
        (0, globals_1.it)("should accept HA mode with multiple origins", () => {
            const haConfig = {
                ...VALID_CONFIG,
                haMode: true,
                wsEndpoint: "wss://ws1.example.com,wss://ws2.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(haConfig)).not.toThrow();
        });
        (0, globals_1.it)("should validate HA mode with single origin without forced logging", () => {
            const consoleSpy = jest.spyOn(console, "warn").mockImplementation();
            const haConfigSingleOrigin = {
                ...VALID_CONFIG,
                haMode: true,
                wsEndpoint: "wss://ws.example.com",
            };
            // Should not throw and should NOT log anything (developers control logging)
            (0, globals_1.expect)(() => (0, client_1.createClient)(haConfigSingleOrigin)).not.toThrow();
            (0, globals_1.expect)(consoleSpy).not.toHaveBeenCalled();
            consoleSpy.mockRestore();
        });
        (0, globals_1.it)("should accept HA mode with origin discovery enabled", () => {
            const haConfigWithDiscovery = {
                ...VALID_CONFIG,
                haMode: true,
                wsEndpoint: "wss://ws.example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(haConfigWithDiscovery)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid HA connection timeout", () => {
            const haConfigWithTimeout = {
                ...VALID_CONFIG,
                haMode: true,
                haConnectionTimeout: 10000,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(haConfigWithTimeout)).not.toThrow();
        });
        (0, globals_1.it)("should validate very low HA connection timeout without forced logging", () => {
            const consoleSpy = jest.spyOn(console, "warn").mockImplementation();
            const haConfigLowTimeout = {
                ...VALID_CONFIG,
                haMode: true,
                haConnectionTimeout: 500, // Less than 1 second
            };
            // Should not throw and should NOT log anything (developers control logging)
            (0, globals_1.expect)(() => (0, client_1.createClient)(haConfigLowTimeout)).not.toThrow();
            (0, globals_1.expect)(consoleSpy).not.toHaveBeenCalled();
            consoleSpy.mockRestore();
        });
        (0, globals_1.it)("should accept connection status callback", () => {
            const callback = jest.fn();
            const haConfigWithCallback = {
                ...VALID_CONFIG,
                haMode: true,
                connectionStatusCallback: callback,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(haConfigWithCallback)).not.toThrow();
        });
    });
    (0, globals_1.describe)("detailed URL validation scenarios", () => {
        (0, globals_1.it)("should reject malformed REST URL with colon prefix", () => {
            const configWithColonUrl = {
                ...VALID_CONFIG,
                endpoint: ":rest.domain.link",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithColonUrl)).toThrow();
        });
        (0, globals_1.it)("should reject malformed WebSocket URL with colon prefix", () => {
            const configWithColonWsUrl = {
                ...VALID_CONFIG,
                wsEndpoint: ":ws.domain.link",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithColonWsUrl)).toThrow();
        });
        (0, globals_1.it)("should reject URLs with invalid characters", () => {
            const configWithInvalidChars = {
                ...VALID_CONFIG,
                endpoint: "https://api[invalid].example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithInvalidChars)).toThrow();
        });
        (0, globals_1.it)("should reject WebSocket URLs with spaces", () => {
            const configWithSpacesInWs = {
                ...VALID_CONFIG,
                wsEndpoint: "wss://ws .example.com",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithSpacesInWs)).toThrow();
        });
        (0, globals_1.it)("should accept valid URLs with subdomains and paths", () => {
            const configWithComplexUrls = {
                ...VALID_CONFIG,
                endpoint: "https://api.prod.dataengine.chain.link/v1/streams",
                wsEndpoint: "wss://ws.prod.dataengine.chain.link/stream/v1",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithComplexUrls)).not.toThrow();
        });
    });
    (0, globals_1.describe)("configuration validation error messages", () => {
        (0, globals_1.it)("should provide clear error message for missing API key", () => {
            const configWithoutApiKey = {
                ...VALID_CONFIG,
                apiKey: undefined,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithoutApiKey)).toThrow(/apiKey/i);
        });
        (0, globals_1.it)("should provide clear error message for missing user secret", () => {
            const configWithoutSecret = {
                ...VALID_CONFIG,
                userSecret: undefined,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithoutSecret)).toThrow(/userSecret/i);
        });
        (0, globals_1.it)("should provide clear error message for invalid endpoint", () => {
            const configWithInvalidEndpoint = {
                ...VALID_CONFIG,
                endpoint: "invalid-url",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithInvalidEndpoint)).toThrow(/endpoint/i);
        });
        (0, globals_1.it)("should provide clear error message for invalid WebSocket endpoint", () => {
            const configWithInvalidWs = {
                ...VALID_CONFIG,
                wsEndpoint: "invalid-ws-url",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithInvalidWs)).toThrow(/websocket|ws/i);
        });
    });
    (0, globals_1.describe)("edge cases and error scenarios", () => {
        (0, globals_1.it)("should reject null configuration", () => {
            (0, globals_1.expect)(() => (0, client_1.createClient)(null)).toThrow();
        });
        (0, globals_1.it)("should reject undefined configuration", () => {
            (0, globals_1.expect)(() => (0, client_1.createClient)(undefined)).toThrow();
        });
        (0, globals_1.it)("should reject empty configuration object", () => {
            (0, globals_1.expect)(() => (0, client_1.createClient)({})).toThrow();
        });
        (0, globals_1.it)("should handle configuration with extra properties", () => {
            const configWithExtra = {
                ...VALID_CONFIG,
                extraProperty: "should be ignored",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithExtra)).not.toThrow();
        });
        (0, globals_1.it)("should reject configuration with wrong type for apiKey", () => {
            const configWithWrongType = {
                ...VALID_CONFIG,
                apiKey: 12345,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithWrongType)).toThrow();
        });
        (0, globals_1.it)("should reject configuration with wrong type for haMode", () => {
            const configWithWrongHaType = {
                ...VALID_CONFIG,
                haMode: "true",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithWrongHaType)).toThrow();
        });
        (0, globals_1.it)("should handle whitespace in URLs", () => {
            const configWithWhitespace = {
                ...VALID_CONFIG,
                endpoint: "  https://api.example.com  ",
                wsEndpoint: "  wss://ws.example.com  ",
            };
            // This might pass or fail depending on our implementation
            // The test documents the current behavior
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithWhitespace)).not.toThrow();
        });
        (0, globals_1.it)("should handle Unicode characters in configuration", () => {
            const configWithUnicode = {
                ...VALID_CONFIG,
                apiKey: "test-api-key-🔑",
                userSecret: "test-secret-🔐",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithUnicode)).not.toThrow();
        });
    });
    (0, globals_1.describe)("configuration defaults", () => {
        (0, globals_1.it)("should apply default values for optional fields", () => {
            const client = (0, client_1.createClient)(VALID_CONFIG);
            // We can't directly inspect the config, but we can test that defaults are used
            (0, globals_1.expect)(client).toBeDefined();
        });
        (0, globals_1.it)("should use default timeout when not specified", () => {
            const client = (0, client_1.createClient)(VALID_CONFIG);
            (0, globals_1.expect)(client).toBeDefined();
            // Default timeout should be applied internally
        });
        (0, globals_1.it)("should use default retry settings when not specified", () => {
            const client = (0, client_1.createClient)(VALID_CONFIG);
            (0, globals_1.expect)(client).toBeDefined();
            // Default retry settings should be applied internally
        });
        (0, globals_1.it)("should use default HA mode settings when not specified", () => {
            const client = (0, client_1.createClient)(VALID_CONFIG);
            (0, globals_1.expect)(client).toBeDefined();
            // HA mode should be disabled by default
        });
        (0, globals_1.it)("should override defaults when explicitly provided", () => {
            const customConfig = {
                ...VALID_CONFIG,
                timeout: 60000,
                retryAttempts: 10,
                haMode: true,
            };
            const client = (0, client_1.createClient)(customConfig);
            (0, globals_1.expect)(client).toBeDefined();
            // Custom values should override defaults
        });
    });
    (0, globals_1.describe)("configuration normalization", () => {
        (0, globals_1.it)("should handle trailing slashes in URLs", () => {
            const configWithTrailingSlashes = {
                ...VALID_CONFIG,
                endpoint: "https://api.example.com/",
                wsEndpoint: "wss://ws.example.com/",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithTrailingSlashes)).not.toThrow();
        });
        (0, globals_1.it)("should handle URLs with paths", () => {
            const configWithPaths = {
                ...VALID_CONFIG,
                endpoint: "https://api.example.com/v1",
                wsEndpoint: "wss://ws.example.com/stream",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithPaths)).not.toThrow();
        });
        (0, globals_1.it)("should handle URLs with query parameters", () => {
            const configWithQuery = {
                ...VALID_CONFIG,
                endpoint: "https://api.example.com?version=1",
                wsEndpoint: "wss://ws.example.com?protocol=v1",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithQuery)).not.toThrow();
        });
        (0, globals_1.it)("should handle URLs with ports", () => {
            const configWithPorts = {
                ...VALID_CONFIG,
                endpoint: "https://api.example.com:8443",
                wsEndpoint: "wss://ws.example.com:8443",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(configWithPorts)).not.toThrow();
        });
    });
    (0, globals_1.describe)("real-world configuration scenarios", () => {
        (0, globals_1.it)("should accept production-like configuration", () => {
            const prodConfig = {
                apiKey: "prod-api-key-abcdef123456",
                userSecret: "prod-secret-xyz789",
                endpoint: "https://api.dataengine.chain.link",
                wsEndpoint: "wss://ws.dataengine.chain.link",
                timeout: 30000,
                retryAttempts: 3,
                retryDelay: 1000,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(prodConfig)).not.toThrow();
        });
        (0, globals_1.it)("should accept testnet configuration", () => {
            const testnetConfig = {
                apiKey: "test-api-key",
                userSecret: "test-secret",
                endpoint: "https://api.testnet-dataengine.chain.link",
                wsEndpoint: "wss://ws.testnet-dataengine.chain.link",
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(testnetConfig)).not.toThrow();
        });
        (0, globals_1.it)("should accept local development configuration", () => {
            const devConfig = {
                apiKey: "dev-key",
                userSecret: "dev-secret",
                endpoint: "http://localhost:3000",
                wsEndpoint: "ws://localhost:3001",
                timeout: 5000,
                retryAttempts: 1,
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(devConfig)).not.toThrow();
        });
        (0, globals_1.it)("should accept HA production configuration", () => {
            const haProdConfig = {
                apiKey: "ha-prod-key",
                userSecret: "ha-prod-secret",
                endpoint: "https://api.dataengine.chain.link",
                wsEndpoint: "wss://ws1.dataengine.chain.link,wss://ws2.dataengine.chain.link",
                haMode: true,
                haConnectionTimeout: 10000,
                connectionStatusCallback: (isConnected, host, origin) => {
                    console.log(`Connection ${isConnected ? "established" : "lost"} to ${host} (${origin})`);
                },
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(haProdConfig)).not.toThrow();
        });
    });
    (0, globals_1.describe)("performance and memory", () => {
        (0, globals_1.it)("should create clients efficiently", () => {
            const start = performance.now();
            // Create 100 clients
            for (let i = 0; i < 100; i++) {
                const client = (0, client_1.createClient)(VALID_CONFIG);
                (0, globals_1.expect)(client).toBeDefined();
            }
            const end = performance.now();
            const duration = end - start;
            // Should complete in reasonable time (less than 1000ms)
            (0, globals_1.expect)(duration).toBeLessThan(1000);
        });
        (0, globals_1.it)("should handle large configuration objects", () => {
            const largeConfig = {
                ...VALID_CONFIG,
                // Add many optional fields
                retryAttempts: 10,
                retryDelay: 2000,
                timeout: 60000,
                haMode: true,
                haConnectionTimeout: 15000,
                connectionStatusCallback: () => { },
                // Simulate large comma-separated URL list
                wsEndpoint: Array(50).fill("wss://ws.example.com").join(","),
            };
            (0, globals_1.expect)(() => (0, client_1.createClient)(largeConfig)).not.toThrow();
        });
    });
});
