// @chainlink/data-streams-sdk/dist/tests/unit/utils/validation/feed-validation.test.js
"use strict";
/**
 * Unit Tests for Feed ID Validation Functions
 *
 * These tests validate the feed ID functionality by:
 * - Testing feed ID hex string parsing and validation
 * - Testing feed ID format validation (length, prefix, characters)
 * - Testing invalid feed ID rejection with clear error messages
 * - Testing feed ID normalization and case handling
 * - Testing version extraction from feed IDs
 * - Testing edge cases (empty, null, malformed, special characters)
 * - Testing feed ID comparison and equality
 * - Testing feed ID array validation
 *
 * Goals:
 * - Ensure robust feed ID validation that prevents invalid data
 * - Test all edge cases and error scenarios comprehensively
 * - Support feed versions V2, V3, V4, V5, V6, V7, V8, V9, V10, V11, V12, V13
 * - Provide clear, helpful error messages for developers
 * - Build the best possible TypeScript feed ID validation
 */
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const validation_1 = require("../../../../src/utils/validation");
const errors_1 = require("../../../../src/types/errors");
(0, globals_1.describe)("Feed ID Validation Tests", () => {
    // Test vectors from reference implementations
    const VALID_FEED_IDS = {
        V1: "0x00016b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V2: "0x00026b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V3: "0x00036b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V4: "0x00046b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V5: "0x00056b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V6: "0x00066b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V7: "0x00076b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V8: "0x00086b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V9: "0x00096b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V10: "0x000a6b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V11: "0x000bfb6d135897e4aaf5657bffd3b0b48f8e2a5131214c9ec2d62eac5d532067",
        V12: "0x000c6b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V13: "0x000d13a9b9c5e37a099f374e92c37914af5c268f3a8a9721f1725135bfb4cbb8",
    };
    const REAL_WORLD_FEED_IDS = {
        // Real feed IDs from reference tests
        FEED1: "0x00020ffa644e6c585a5bec0e25ca476b6666666666e22b6240957720dcba0e14",
        FEED2: "0x00020ffa644e6c585a88888825ca476b6666666666e22b6240957720dcba0e14",
    };
    (0, globals_1.describe)("valid feed ID formats", () => {
        (0, globals_1.it)("should reject unsupported V1 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V1)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V1)).toThrow("Invalid feed ID version");
        });
        (0, globals_1.it)("should accept valid V2 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V2)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V3 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V3)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V4 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V4)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V5 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V5)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V6 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V6)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V7 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V7)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V8 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V8)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V9 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V9)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V10 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V10)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V11 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V11)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V12 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V12)).not.toThrow();
        });
        (0, globals_1.it)("should accept valid V13 feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V13)).not.toThrow();
        });
        (0, globals_1.it)("should accept real-world feed IDs", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(REAL_WORLD_FEED_IDS.FEED1)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(REAL_WORLD_FEED_IDS.FEED2)).not.toThrow();
        });
        (0, globals_1.it)("should reject uppercase prefix but accept uppercase hex", () => {
            const uppercaseFeedId = VALID_FEED_IDS.V3.toUpperCase(); // "0X..." format
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(uppercaseFeedId)).toThrow(errors_1.ValidationError);
            // But lowercase prefix with uppercase hex should work
            const mixedCase = "0x" + VALID_FEED_IDS.V3.slice(2).toUpperCase();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(mixedCase)).not.toThrow();
        });
        (0, globals_1.it)("should accept mixed case hex characters", () => {
            const mixedCaseFeedId = "0x00036B4aa7E57ca7B68ae1BF45653f56B656fd3AA335ef7fAE696b663F1b8472";
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(mixedCaseFeedId)).not.toThrow();
        });
    });
    (0, globals_1.describe)("feed ID format validation", () => {
        (0, globals_1.it)("should reject feed ID without 0x prefix", () => {
            const withoutPrefix = VALID_FEED_IDS.V3.slice(2); // Remove 0x
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(withoutPrefix)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(withoutPrefix)).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject feed ID with wrong prefix", () => {
            const wrongPrefix = "0y" + VALID_FEED_IDS.V3.slice(2);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(wrongPrefix)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(wrongPrefix)).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject feed ID with incorrect length (too short)", () => {
            const tooShort = "0x00036b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b84"; // 62 chars instead of 64
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(tooShort)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(tooShort)).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject feed ID with incorrect length (too long)", () => {
            const tooLong = VALID_FEED_IDS.V3 + "72"; // Extra characters
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(tooLong)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(tooLong)).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject feed ID with invalid hex characters", () => {
            const invalidHex = "0x00036g4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472"; // 'g' is not valid hex
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(invalidHex)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(invalidHex)).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject feed ID with special characters", () => {
            const specialChars = "0x00036b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b84@#";
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(specialChars)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(specialChars)).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject feed ID with spaces", () => {
            const withSpaces = "0x0003 6b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472";
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(withSpaces)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(withSpaces)).toThrow("Invalid feed ID format");
        });
    });
    (0, globals_1.describe)("feed ID version validation", () => {
        (0, globals_1.it)("should reject unsupported version V1 (0x0001)", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V1)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V1)).toThrow("Invalid feed ID version");
        });
        (0, globals_1.it)("should accept supported version V2 (0x0002)", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V2)).not.toThrow();
        });
        (0, globals_1.it)("should accept supported version V3 (0x0003)", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V3)).not.toThrow();
        });
        (0, globals_1.it)("should accept supported version V4 (0x0004)", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V4)).not.toThrow();
        });
        (0, globals_1.it)("should accept supported version V5 (0x0005)", () => {
            const v5FeedId = "0x00056b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472";
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(v5FeedId)).not.toThrow();
        });
        (0, globals_1.it)("should reject version 0 (0x0000)", () => {
            const v0FeedId = "0x00006b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472";
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(v0FeedId)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(v0FeedId)).toThrow("Invalid feed ID version");
        });
        (0, globals_1.it)("should reject very high version numbers", () => {
            const highVersionFeedId = "0xFFFF6b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472";
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(highVersionFeedId)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(highVersionFeedId)).toThrow("Invalid feed ID version");
        });
    });
    (0, globals_1.describe)("edge cases and error scenarios", () => {
        (0, globals_1.it)("should reject empty string", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)("")).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)("")).toThrow("Feed ID is required");
        });
        (0, globals_1.it)("should reject null feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(null)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(null)).toThrow("Feed ID is required");
        });
        (0, globals_1.it)("should reject undefined feed ID", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(undefined)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(undefined)).toThrow("Feed ID is required");
        });
        (0, globals_1.it)("should reject whitespace-only string", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)("   ")).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)("   ")).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject number instead of string", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(123)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(123)).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject object instead of string", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)({})).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)({})).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject array instead of string", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)([])).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)([])).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject boolean instead of string", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(true)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(true)).toThrow("Invalid feed ID format");
        });
    });
    (0, globals_1.describe)("feed ID normalization and case handling", () => {
        (0, globals_1.it)("should handle leading/trailing whitespace (if we decide to be permissive)", () => {
            // Note: Our current implementation doesn't trim, but we could enhance it
            const withWhitespace = `  ${VALID_FEED_IDS.V3}  `;
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(withWhitespace)).toThrow(); // Current behavior
        });
        (0, globals_1.it)("should be case insensitive for hex characters but not prefix", () => {
            const lowercase = VALID_FEED_IDS.V3.toLowerCase();
            const uppercaseHex = "0x" + VALID_FEED_IDS.V3.slice(2).toUpperCase();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(lowercase)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(uppercaseHex)).not.toThrow();
        });
        (0, globals_1.it)("should reject uppercase 0X prefix", () => {
            const uppercasePrefix = VALID_FEED_IDS.V3.replace("0x", "0X");
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(uppercasePrefix)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(uppercasePrefix)).toThrow("Invalid feed ID format");
        });
    });
    (0, globals_1.describe)("feed ID array validation", () => {
        (0, globals_1.it)("should accept array of valid feed IDs", () => {
            const validArray = [VALID_FEED_IDS.V2, VALID_FEED_IDS.V3, VALID_FEED_IDS.V4];
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(validArray)).not.toThrow();
        });
        (0, globals_1.it)("should accept single feed ID in array", () => {
            const singleArray = [VALID_FEED_IDS.V3];
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(singleArray)).not.toThrow();
        });
        (0, globals_1.it)("should reject empty array", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)([])).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)([])).toThrow("At least one feed ID is required");
        });
        (0, globals_1.it)("should reject non-array input", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)("not-an-array")).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)("not-an-array")).toThrow("Feed IDs must be an array");
        });
        (0, globals_1.it)("should reject null array", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(null)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(null)).toThrow("Feed IDs must be an array");
        });
        (0, globals_1.it)("should reject undefined array", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(undefined)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(undefined)).toThrow("Feed IDs must be an array");
        });
        (0, globals_1.it)("should reject array with invalid feed ID", () => {
            const mixedArray = [VALID_FEED_IDS.V3, "invalid-feed-id", VALID_FEED_IDS.V4];
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(mixedArray)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(mixedArray)).toThrow("Invalid feed ID format");
        });
        (0, globals_1.it)("should reject array with empty string", () => {
            const arrayWithEmpty = [VALID_FEED_IDS.V3, "", VALID_FEED_IDS.V4];
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(arrayWithEmpty)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(arrayWithEmpty)).toThrow("Feed ID is required");
        });
        (0, globals_1.it)("should reject array with null element", () => {
            const arrayWithNull = [VALID_FEED_IDS.V3, null, VALID_FEED_IDS.V4];
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(arrayWithNull)).toThrow(errors_1.ValidationError);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(arrayWithNull)).toThrow("Feed ID is required");
        });
        (0, globals_1.it)("should handle large arrays efficiently", () => {
            const largeArray = Array(1000).fill(VALID_FEED_IDS.V3);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(largeArray)).not.toThrow();
        });
        (0, globals_1.it)("should handle duplicate feed IDs in array", () => {
            const duplicateArray = [VALID_FEED_IDS.V3, VALID_FEED_IDS.V3, VALID_FEED_IDS.V4];
            (0, globals_1.expect)(() => (0, validation_1.validateFeedIds)(duplicateArray)).not.toThrow(); // Duplicates are allowed
        });
    });
    (0, globals_1.describe)("error message quality", () => {
        (0, globals_1.it)("should provide specific error for missing prefix", () => {
            const withoutPrefix = VALID_FEED_IDS.V3.slice(2);
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(withoutPrefix)).toThrow("Invalid feed ID format. Must be 0x followed by 64 hex characters");
        });
        (0, globals_1.it)("should provide specific error for wrong length", () => {
            const wrongLength = "0x123";
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(wrongLength)).toThrow("Invalid feed ID format. Must be 0x followed by 64 hex characters");
        });
        (0, globals_1.it)("should provide specific error for unsupported version", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V1)).toThrow("Invalid feed ID version. Must start with 0x0002, 0x0003, 0x0004, 0x0005, 0x0006, 0x0007, 0x0008, 0x0009, 0x000a, 0x000b, 0x000c or 0x000d");
        });
        (0, globals_1.it)("should provide helpful error for empty input", () => {
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)("")).toThrow("Feed ID is required");
        });
    });
    (0, globals_1.describe)("performance and efficiency", () => {
        (0, globals_1.it)("should validate feed IDs efficiently", () => {
            const start = performance.now();
            // Validate 1000 feed IDs
            for (let i = 0; i < 1000; i++) {
                (0, validation_1.validateFeedId)(VALID_FEED_IDS.V3);
            }
            const end = performance.now();
            const duration = end - start;
            // Should complete in reasonable time (less than 100ms)
            (0, globals_1.expect)(duration).toBeLessThan(100);
        });
        (0, globals_1.it)("should fail fast for obviously invalid inputs", () => {
            const start = performance.now();
            try {
                (0, validation_1.validateFeedId)("");
            }
            catch {
                // Expected to throw
            }
            const end = performance.now();
            const duration = end - start;
            // Should fail very quickly (less than 1ms)
            (0, globals_1.expect)(duration).toBeLessThan(1);
        });
    });
    (0, globals_1.describe)("integration with validation constants", () => {
        (0, globals_1.it)("should use the same regex patterns as constants", () => {
            // This test ensures our validation logic is consistent
            // We're testing the behavior rather than implementation details
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V3)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)("invalid")).toThrow();
        });
        (0, globals_1.it)("should support all documented feed versions", () => {
            // Test that our version validation matches what we claim to support
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V2)).not.toThrow(); // V2 supported
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V3)).not.toThrow(); // V3 supported
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V4)).not.toThrow(); // V4 supported
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V5)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V6)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V7)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V8)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V9)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V10)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V11)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V12)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V13)).not.toThrow();
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(VALID_FEED_IDS.V1)).toThrow(); // V1 not supported in our implementation
        });
    });
    (0, globals_1.describe)("real-world compatibility", () => {
        (0, globals_1.it)("should accept feed IDs from examples", () => {
            // These are actual feed IDs used in tests
            const goSdkFeedIds = [
                "0x00020ffa644e6c585a5bec0e25ca476b6666666666e22b6240957720dcba0e14",
                "0x00020ffa644e6c585a88888825ca476b6666666666e22b6240957720dcba0e14",
            ];
            goSdkFeedIds.forEach(feedId => {
                (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(feedId)).not.toThrow();
            });
        });
        (0, globals_1.it)("should accept feed IDs from additional examples", () => {
            // These are additional test vectors
            const rustSdkFeedIds = [
                "0x00016b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472", // V1 (we reject)
                "0x00026b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472", // V2 (we accept)
                "0x00036b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472", // V3 (we accept)
                "0x00046b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472", // V4 (we accept)
            ];
            // V1 is not supported in our implementation (we only support V2, V3, V4)
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(rustSdkFeedIds[0])).toThrow(); // V1
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(rustSdkFeedIds[1])).not.toThrow(); // V2
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(rustSdkFeedIds[2])).not.toThrow(); // V3
            (0, globals_1.expect)(() => (0, validation_1.validateFeedId)(rustSdkFeedIds[3])).not.toThrow(); // V4
        });
    });
});
