// @chainlink/data-streams-sdk/dist/tests/unit/utils/validation/report-validation.test.js
"use strict";
/**
 * Unit Tests for Report Validation and Decoding
 *
 * These tests validate the report functionality by:
 * - Testing report structure validation for all versions (V2, V3, V4, V5, V6, V7, V8, V9, V10, V11, V12, V13)
 * - Testing report version handling and extraction
 * - Testing malformed report rejection with clear error messages
 * - Testing report timestamp validation
 * - Testing report data integrity checks
 * - Testing report metadata extraction
 * - Testing ABI decoding edge cases
 * - Testing market status validation (V4, V8)
 * - Testing ripcord validation (V9)
 * - Testing large number handling (int192, uint192, uint64)
 *
 * Goals:
 * - Ensure robust report validation that prevents invalid data
 * - Test all edge cases and error scenarios comprehensively
 * - Support all report versions (V2, V3, V4, V5, V6, V7, V8, V9, V10, V11, V12, V13)
 * - Provide clear, helpful error messages for developers
 * - Build the best possible TypeScript report validation
 */
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const decoder_1 = require("../../../../src/decoder");
const errors_1 = require("../../../../src/types/errors");
const types_1 = require("../../../../src/types");
const ethers_1 = require("ethers");
(0, globals_1.describe)("Report Validation Tests", () => {
    // Valid feed IDs for different versions
    const FEED_IDS = {
        V2: "0x00026b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V3: "0x00036b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V4: "0x00046b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V5: "0x00056b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V6: "0x00066b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V7: "0x00076b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V8: "0x00086b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V9: "0x00096b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V10: "0x000a6b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V11: "0x000bfb6d135897e4aaf5657bffd3b0b48f8e2a5131214c9ec2d62eac5d532067",
        V12: "0x000c6b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472",
        V13: "0x000d13a9b9c5e37a099f374e92c37914af5c268f3a8a9721f1725135bfb4cbb8",
    };
    // Helper function to create a valid full report structure
    function createFullReport(reportBlob) {
        const abiCoder = new ethers_1.AbiCoder();
        // Create mock report context (3 x bytes32)
        const reportContext = [
            "0x0001020304050607080910111213141516171819202122232425262728293031",
            "0x3132333435363738394041424344454647484950515253545556575859606162",
            "0x6364656667686970717273747576777879808182838485868788899091929394",
        ];
        // Create mock signature data
        const rawRs = [
            "0x1111111111111111111111111111111111111111111111111111111111111111",
            "0x2222222222222222222222222222222222222222222222222222222222222222",
        ];
        const rawSs = [
            "0x3333333333333333333333333333333333333333333333333333333333333333",
            "0x4444444444444444444444444444444444444444444444444444444444444444",
        ];
        const rawVs = "0x5555555555555555555555555555555555555555555555555555555555555555";
        return abiCoder.encode(["bytes32[3]", "bytes", "bytes32[]", "bytes32[]", "bytes32"], [reportContext, reportBlob, rawRs, rawSs, rawVs]);
    }
    // Helper function to create a valid V2 report blob
    function createV2ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192"], [
            FEED_IDS.V2, // feedId
            1640995200, // validFromTimestamp
            1640995300, // observationsTimestamp
            "1000000000000000000", // nativeFee (1 ETH in wei)
            "500000000000000000", // linkFee (0.5 LINK)
            1640995400, // expiresAt
            "2000000000000000000000", // price (2000 USD with 18 decimals)
        ]);
    }
    // Helper function to create a valid V3 report blob
    function createV3ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "int192", "int192"], [
            FEED_IDS.V3, // feedId
            1640995200, // validFromTimestamp
            1640995300, // observationsTimestamp
            "1000000000000000000", // nativeFee
            "500000000000000000", // linkFee
            1640995400, // expiresAt
            "2000000000000000000000", // price (benchmark)
            "1995000000000000000000", // bid
            "2005000000000000000000", // ask
        ]);
    }
    // Helper function to create a valid V4 report blob
    function createV4ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint8"], [
            FEED_IDS.V4, // feedId
            1640995200, // validFromTimestamp
            1640995300, // observationsTimestamp
            "1000000000000000000", // nativeFee
            "500000000000000000", // linkFee
            1640995400, // expiresAt
            "2000000000000000000000", // price
            types_1.MarketStatus.ACTIVE, // marketStatus
        ]);
    }
    // Helper function to create a valid V8 report blob
    function createV8ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "uint64", "int192", "uint32"], [
            FEED_IDS.V8, // feedId
            1640995200, // validFromTimestamp
            1640995300, // observationsTimestamp
            "1000000000000000000", // nativeFee
            "500000000000000000", // linkFee
            1640995400, // expiresAt
            1640995250, // lastUpdateTimestamp
            "2500000000000000000000", // midPrice (2500 USD with 18 decimals)
            types_1.MarketStatus.ACTIVE, // marketStatus
        ]);
    }
    // Helper function to create a valid V9 report blob
    function createV9ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint64", "int192", "uint32"], [
            FEED_IDS.V9, // feedId
            1640995200, // validFromTimestamp
            1640995300, // observationsTimestamp
            "1000000000000000000", // nativeFee
            "500000000000000000", // linkFee
            1640995400, // expiresAt
            "1050000000000000000", // navPerShare ($1.05 with 18 decimals)
            1640995250, // navDate
            "100000000000000000000000", // aum ($100k with 18 decimals)
            0, // ripcord (normal)
        ]);
    }
    // Helper function to create a valid V5 report blob
    function createV5ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint32", "uint32"], [
            FEED_IDS.V5,
            1640995200,
            1640995300,
            "1000000000000000000",
            "500000000000000000",
            1640995400,
            "1234567890000000000",
            1640995250,
            3600,
        ]);
    }
    // Helper function to create a valid V6 report blob
    function createV6ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "int192", "int192", "int192", "int192"], [
            FEED_IDS.V6,
            1640995200,
            1640995300,
            "1000000000000000000",
            "500000000000000000",
            1640995400,
            "2000000000000000000000",
            "2100000000000000000000",
            "2200000000000000000000",
            "2300000000000000000000",
            "2400000000000000000000",
        ]);
    }
    // Helper function to create a valid V7 report blob
    function createV7ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192"], [
            FEED_IDS.V7,
            1640995200,
            1640995300,
            "1000000000000000000",
            "500000000000000000",
            1640995400,
            "987654321000000000",
        ]);
    }
    // Helper function to create a valid V10 report blob
    function createV10ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode([
            "bytes32",
            "uint32",
            "uint32",
            "uint192",
            "uint192",
            "uint32",
            "uint64",
            "int192",
            "uint32",
            "int192",
            "int192",
            "uint32",
            "int192",
        ], [
            FEED_IDS.V10,
            1640995200,
            1640995300,
            "1000000000000000000",
            "500000000000000000",
            1640995400,
            1640995250,
            "75000000000000000000",
            types_1.MarketStatus.ACTIVE,
            "1000000000000000000",
            "1100000000000000000",
            1641081600,
            "150000000000000000000",
        ]);
    }
    // Helper function to create a valid V11 report blob
    function createV11ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode([
            "bytes32",
            "uint32",
            "uint32",
            "uint192",
            "uint192",
            "uint32",
            "int192",
            "uint64",
            "int192",
            "int192",
            "int192",
            "int192",
            "int192",
            "uint32",
        ], [
            FEED_IDS.V11,
            1640995200,
            1640995300,
            "1000000000000000000",
            "500000000000000000",
            1640995400,
            "103000000000000000000",
            1640995250000000,
            "101000000000000000000",
            "10001",
            "105000000000000000000",
            "10002",
            "103000000000000000000",
            types_1.MarketStatus.ACTIVE,
        ]);
    }
    // Helper function to create a valid V12 report blob
    function createV12ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode([
            "bytes32",
            "uint32",
            "uint32",
            "uint192",
            "uint192",
            "uint32",
            "int192",
            "int192",
            "uint64",
            "uint32",
        ], [
            FEED_IDS.V12,
            1640995200,
            1640995300,
            "1000000000000000000",
            "500000000000000000",
            1640995400,
            "103000000000000000000",
            "104000000000000000000",
            1640995250000000,
            types_1.Ripcord.NORMAL,
        ]);
    }
    // Helper function to create a valid V13 report blob
    function createV13ReportBlob() {
        const abiCoder = new ethers_1.AbiCoder();
        return abiCoder.encode([
            "bytes32",
            "uint32",
            "uint32",
            "uint192",
            "uint192",
            "uint32",
            "int192",
            "int192",
            "uint64",
            "uint64",
            "int192",
        ], [
            FEED_IDS.V13,
            1640995200,
            1640995300,
            "1000000000000000000",
            "500000000000000000",
            1640995400,
            "75000000000000000000",
            "78000000000000000000",
            10000,
            11000,
            "76000000000000000000",
        ]);
    }
    (0, globals_1.describe)("valid report decoding", () => {
        (0, globals_1.it)("should decode valid V2 report", () => {
            const reportBlob = createV2ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            (0, globals_1.expect)(decoded.version).toBe("V2");
            (0, globals_1.expect)(decoded.nativeFee).toBe(1000000000000000000n);
            (0, globals_1.expect)(decoded.linkFee).toBe(500000000000000000n);
            (0, globals_1.expect)(decoded.expiresAt).toBe(1640995400);
            (0, globals_1.expect)(decoded.price).toBe(2000000000000000000000n);
        });
        (0, globals_1.it)("should decode valid V3 report", () => {
            const reportBlob = createV3ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V3);
            (0, globals_1.expect)(decoded.version).toBe("V3");
            (0, globals_1.expect)(decoded.nativeFee).toBe(1000000000000000000n);
            (0, globals_1.expect)(decoded.linkFee).toBe(500000000000000000n);
            (0, globals_1.expect)(decoded.expiresAt).toBe(1640995400);
            (0, globals_1.expect)(decoded.price).toBe(2000000000000000000000n);
            (0, globals_1.expect)(decoded.bid).toBe(1995000000000000000000n);
            (0, globals_1.expect)(decoded.ask).toBe(2005000000000000000000n);
        });
        (0, globals_1.it)("should decode valid V4 report", () => {
            const reportBlob = createV4ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V4);
            (0, globals_1.expect)(decoded.version).toBe("V4");
            (0, globals_1.expect)(decoded.nativeFee).toBe(1000000000000000000n);
            (0, globals_1.expect)(decoded.linkFee).toBe(500000000000000000n);
            (0, globals_1.expect)(decoded.expiresAt).toBe(1640995400);
            (0, globals_1.expect)(decoded.price).toBe(2000000000000000000000n);
            (0, globals_1.expect)(decoded.marketStatus).toBe(types_1.MarketStatus.ACTIVE);
        });
        (0, globals_1.it)("should decode valid V8 report", () => {
            const reportBlob = createV8ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V8);
            (0, globals_1.expect)(decoded.version).toBe("V8");
            (0, globals_1.expect)(decoded.nativeFee).toBe(1000000000000000000n);
            (0, globals_1.expect)(decoded.linkFee).toBe(500000000000000000n);
            (0, globals_1.expect)(decoded.expiresAt).toBe(1640995400);
            (0, globals_1.expect)(decoded.midPrice).toBe(2500000000000000000000n);
            (0, globals_1.expect)(decoded.lastUpdateTimestamp).toBe(1640995250);
            (0, globals_1.expect)(decoded.marketStatus).toBe(types_1.MarketStatus.ACTIVE);
        });
        (0, globals_1.it)("should decode valid V9 report", () => {
            const reportBlob = createV9ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V9);
            (0, globals_1.expect)(decoded.version).toBe("V9");
            (0, globals_1.expect)(decoded.nativeFee).toBe(1000000000000000000n);
            (0, globals_1.expect)(decoded.linkFee).toBe(500000000000000000n);
            (0, globals_1.expect)(decoded.expiresAt).toBe(1640995400);
            (0, globals_1.expect)(decoded.navPerShare).toBe(1050000000000000000n);
            (0, globals_1.expect)(decoded.navDate).toBe(1640995250);
            (0, globals_1.expect)(decoded.aum).toBe(100000000000000000000000n);
            (0, globals_1.expect)(decoded.ripcord).toBe(0);
        });
        (0, globals_1.it)("should decode valid V10 report", () => {
            const reportBlob = createV10ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V10);
            (0, globals_1.expect)(decoded.version).toBe("V10");
            (0, globals_1.expect)(decoded.price).toBe(75000000000000000000n);
            (0, globals_1.expect)(decoded.tokenizedPrice).toBe(150000000000000000000n);
        });
        (0, globals_1.it)("should decode valid V11 report", () => {
            const reportBlob = createV11ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V11);
            (0, globals_1.expect)(decoded.version).toBe("V11");
            (0, globals_1.expect)(decoded.mid).toBe(103000000000000000000n);
            (0, globals_1.expect)(decoded.bid).toBe(101000000000000000000n);
            (0, globals_1.expect)(decoded.ask).toBe(105000000000000000000n);
            (0, globals_1.expect)(decoded.lastTradedPrice).toBe(103000000000000000000n);
        });
        (0, globals_1.it)("should decode valid V12 report", () => {
            const reportBlob = createV12ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V12);
            (0, globals_1.expect)(decoded.version).toBe("V12");
            (0, globals_1.expect)(decoded.nativeFee).toBe(1000000000000000000n);
            (0, globals_1.expect)(decoded.linkFee).toBe(500000000000000000n);
            (0, globals_1.expect)(decoded.expiresAt).toBe(1640995400);
            (0, globals_1.expect)(decoded.navPerShare).toBe(103000000000000000000n);
            (0, globals_1.expect)(decoded.nextNavPerShare).toBe(104000000000000000000n);
            (0, globals_1.expect)(decoded.navDate).toBe(1640995250000000);
            (0, globals_1.expect)(decoded.ripcord).toBe(0);
        });
        (0, globals_1.it)("should decode valid V13 report", () => {
            const reportBlob = createV13ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V13);
            (0, globals_1.expect)(decoded.version).toBe("V13");
            (0, globals_1.expect)(decoded.bestAsk).toBe(75000000000000000000n);
            (0, globals_1.expect)(decoded.bestBid).toBe(78000000000000000000n);
            (0, globals_1.expect)(decoded.lastTradedPrice).toBe(76000000000000000000n);
        });
        (0, globals_1.it)("should handle reports without 0x prefix", () => {
            const reportBlob = createV2ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const withoutPrefix = fullReport.slice(2); // Remove 0x
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(withoutPrefix, FEED_IDS.V2)).toThrow(errors_1.ReportDecodingError);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(withoutPrefix, FEED_IDS.V2)).toThrow("Report hex string must start with 0x");
        });
        (0, globals_1.it)("should handle large price values", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const largePrice = "999999999999999999999999999999999999999999999999"; // Very large int192
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192"], [FEED_IDS.V2, 1640995200, 1640995300, "1000000000000000000", "500000000000000000", 1640995400, largePrice]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            (0, globals_1.expect)(decoded.price).toBe(BigInt(largePrice));
        });
        (0, globals_1.it)("should handle negative price values", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const negativePrice = "-1000000000000000000000"; // Negative price
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192"], [FEED_IDS.V2, 1640995200, 1640995300, "1000000000000000000", "500000000000000000", 1640995400, negativePrice]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            (0, globals_1.expect)(decoded.price).toBe(BigInt(negativePrice));
        });
        (0, globals_1.it)("should handle zero values", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192"], [
                FEED_IDS.V2,
                0, // validFromTimestamp
                0, // observationsTimestamp
                "0", // nativeFee
                "0", // linkFee
                0, // expiresAt
                "0", // price
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            (0, globals_1.expect)(decoded.nativeFee).toBe(0n);
            (0, globals_1.expect)(decoded.linkFee).toBe(0n);
            (0, globals_1.expect)(decoded.expiresAt).toBe(0);
            (0, globals_1.expect)(decoded.price).toBe(0n);
        });
    });
    (0, globals_1.describe)("report version handling", () => {
        (0, globals_1.it)("should reject unsupported V1 version", () => {
            const v1FeedId = "0x00016b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472";
            const reportBlob = createV2ReportBlob(); // Use V2 blob but V1 feed ID
            const fullReport = createFullReport(reportBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, v1FeedId)).toThrow(errors_1.ReportDecodingError);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, v1FeedId)).toThrow("Unknown report version: 0x0001");
        });
        (0, globals_1.it)("should reject unknown version V0", () => {
            const v0FeedId = "0x00006b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472";
            const reportBlob = createV2ReportBlob();
            const fullReport = createFullReport(reportBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, v0FeedId)).toThrow(errors_1.ReportDecodingError);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, v0FeedId)).toThrow("Unknown report version: 0x0000");
        });
        (0, globals_1.it)("should reject version 0", () => {
            const v0FeedId = "0x00006b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472";
            const reportBlob = createV2ReportBlob();
            const fullReport = createFullReport(reportBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, v0FeedId)).toThrow(errors_1.ReportDecodingError);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, v0FeedId)).toThrow("Unknown report version: 0x0000");
        });
        (0, globals_1.it)("should extract version from feed ID correctly", () => {
            // Test that version extraction works for all supported versions
            const versions = [
                { feedId: FEED_IDS.V2, expectedVersion: "V2" },
                { feedId: FEED_IDS.V3, expectedVersion: "V3" },
                { feedId: FEED_IDS.V4, expectedVersion: "V4" },
                { feedId: FEED_IDS.V5, expectedVersion: "V5" },
                { feedId: FEED_IDS.V6, expectedVersion: "V6" },
                { feedId: FEED_IDS.V7, expectedVersion: "V7" },
                { feedId: FEED_IDS.V8, expectedVersion: "V8" },
                { feedId: FEED_IDS.V9, expectedVersion: "V9" },
                { feedId: FEED_IDS.V10, expectedVersion: "V10" },
                { feedId: FEED_IDS.V11, expectedVersion: "V11" },
                { feedId: FEED_IDS.V12, expectedVersion: "V12" },
                { feedId: FEED_IDS.V13, expectedVersion: "V13" },
            ];
            versions.forEach(({ feedId, expectedVersion }) => {
                const reportBlob = expectedVersion === "V2"
                    ? createV2ReportBlob()
                    : expectedVersion === "V3"
                        ? createV3ReportBlob()
                        : expectedVersion === "V4"
                            ? createV4ReportBlob()
                            : expectedVersion === "V5"
                                ? createV5ReportBlob()
                                : expectedVersion === "V6"
                                    ? createV6ReportBlob()
                                    : expectedVersion === "V7"
                                        ? createV7ReportBlob()
                                        : expectedVersion === "V8"
                                            ? createV8ReportBlob()
                                            : expectedVersion === "V9"
                                                ? createV9ReportBlob()
                                                : expectedVersion === "V10"
                                                    ? createV10ReportBlob()
                                                    : expectedVersion === "V11"
                                                        ? createV11ReportBlob()
                                                        : expectedVersion === "V12"
                                                            ? createV12ReportBlob()
                                                            : createV13ReportBlob();
                const fullReport = createFullReport(reportBlob);
                const decoded = (0, decoder_1.decodeReport)(fullReport, feedId);
                (0, globals_1.expect)(decoded.version).toBe(expectedVersion);
            });
        });
    });
    (0, globals_1.describe)("malformed report rejection", () => {
        (0, globals_1.it)("should reject empty report", () => {
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)("", FEED_IDS.V2)).toThrow(errors_1.ReportDecodingError);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)("", FEED_IDS.V2)).toThrow("Report hex string must start with 0x");
        });
        (0, globals_1.it)("should reject invalid hex string", () => {
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)("not-hex-string", FEED_IDS.V2)).toThrow(errors_1.ReportDecodingError);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)("not-hex-string", FEED_IDS.V2)).toThrow("Report hex string must start with 0x");
        });
        (0, globals_1.it)("should reject malformed hex data", () => {
            const invalidHex = "0xZZZZZZZZ"; // Invalid hex characters
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(invalidHex, FEED_IDS.V2)).toThrow(errors_1.ReportDecodingError);
        });
        (0, globals_1.it)("should reject truncated report data", () => {
            const reportBlob = createV2ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const truncated = fullReport.slice(0, 100); // Truncate the report
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(truncated, FEED_IDS.V2)).toThrow(errors_1.ReportDecodingError);
        });
        (0, globals_1.it)("should reject report with wrong structure", () => {
            const abiCoder = new ethers_1.AbiCoder();
            // Create malformed structure (missing fields)
            const malformedReport = abiCoder.encode(["bytes32", "uint32"], // Only 2 fields instead of required structure
            [FEED_IDS.V2, 1640995200]);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(malformedReport, FEED_IDS.V2)).toThrow(errors_1.ReportDecodingError);
        });
        (0, globals_1.it)("should reject report blob with insufficient data for V2", () => {
            const abiCoder = new ethers_1.AbiCoder();
            // Create report blob with insufficient fields for V2 (needs 7 fields)
            const insufficientBlob = abiCoder.encode(["bytes32", "uint32", "uint32"], // Only 3 fields
            [FEED_IDS.V2, 1640995200, 1640995300]);
            const fullReport = createFullReport(insufficientBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2)).toThrow(errors_1.ReportDecodingError);
        });
        (0, globals_1.it)("should reject report blob with insufficient data for V3", () => {
            const abiCoder = new ethers_1.AbiCoder();
            // Create report blob with insufficient fields for V3 (needs 9 fields)
            const insufficientBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192"], // Only 5 fields
            [FEED_IDS.V3, 1640995200, 1640995300, "1000000000000000000", "500000000000000000"]);
            const fullReport = createFullReport(insufficientBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V3)).toThrow(errors_1.ReportDecodingError);
        });
        (0, globals_1.it)("should reject report blob with insufficient data for V4", () => {
            const abiCoder = new ethers_1.AbiCoder();
            // Create report blob with insufficient fields for V4 (needs 8 fields)
            const insufficientBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192"], // Only 4 fields
            [FEED_IDS.V4, 1640995200, 1640995300, "1000000000000000000"]);
            const fullReport = createFullReport(insufficientBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V4)).toThrow(errors_1.ReportDecodingError);
        });
    });
    (0, globals_1.describe)("market status validation (V4)", () => {
        (0, globals_1.it)("should accept valid ACTIVE market status", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint8"], [
                FEED_IDS.V4,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                "2000000000000000000000",
                types_1.MarketStatus.ACTIVE,
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V4);
            (0, globals_1.expect)(decoded.marketStatus).toBe(types_1.MarketStatus.ACTIVE);
        });
        (0, globals_1.it)("should accept valid INACTIVE market status", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint8"], [
                FEED_IDS.V4,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                "2000000000000000000000",
                types_1.MarketStatus.INACTIVE,
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V4);
            (0, globals_1.expect)(decoded.marketStatus).toBe(types_1.MarketStatus.INACTIVE);
        });
        (0, globals_1.it)("should reject invalid market status", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const invalidMarketStatus = 99; // Invalid status
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint8"], [
                FEED_IDS.V4,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                "2000000000000000000000",
                invalidMarketStatus,
            ]);
            const fullReport = createFullReport(reportBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V4)).toThrow(errors_1.ReportDecodingError);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V4)).toThrow("Invalid market status: 99");
        });
    });
    (0, globals_1.describe)("market status validation (V8)", () => {
        (0, globals_1.it)("should accept valid ACTIVE market status", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "uint64", "int192", "uint32"], [
                FEED_IDS.V8,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                1640995250,
                "2500000000000000000000",
                types_1.MarketStatus.ACTIVE,
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V8);
            (0, globals_1.expect)(decoded.marketStatus).toBe(types_1.MarketStatus.ACTIVE);
        });
        (0, globals_1.it)("should reject invalid market status", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const invalidMarketStatus = 99; // Invalid status
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "uint64", "int192", "uint32"], [
                FEED_IDS.V8,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                1640995250,
                "2500000000000000000000",
                invalidMarketStatus,
            ]);
            const fullReport = createFullReport(reportBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V8)).toThrow(errors_1.ReportDecodingError);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V8)).toThrow("Invalid market status: 99");
        });
    });
    (0, globals_1.describe)("ripcord validation (V9)", () => {
        (0, globals_1.it)("should accept normal ripcord (0)", () => {
            const reportBlob = createV9ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V9);
            (0, globals_1.expect)(decoded.ripcord).toBe(0);
        });
        (0, globals_1.it)("should accept paused ripcord (1)", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint64", "int192", "uint32"], [
                FEED_IDS.V9,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                "1050000000000000000",
                1640995250,
                "100000000000000000000000",
                1, // Paused ripcord
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V9);
            (0, globals_1.expect)(decoded.ripcord).toBe(1);
        });
    });
    (0, globals_1.describe)("edge cases and boundary conditions", () => {
        (0, globals_1.it)("should handle maximum uint32 values", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const maxUint32 = "4294967295"; // 2^32 - 1
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192"], [
                FEED_IDS.V2,
                maxUint32, // validFromTimestamp
                maxUint32, // observationsTimestamp
                "1000000000000000000",
                "500000000000000000",
                maxUint32, // expiresAt
                "2000000000000000000000",
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            (0, globals_1.expect)(decoded.expiresAt).toBe(4294967295);
        });
        (0, globals_1.it)("should handle maximum uint192 values", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const maxUint192 = "6277101735386680763835789423207666416102355444464034512895"; // 2^192 - 1
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192"], [
                FEED_IDS.V2,
                1640995200,
                1640995300,
                maxUint192, // nativeFee
                maxUint192, // linkFee
                1640995400,
                "2000000000000000000000",
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            (0, globals_1.expect)(decoded.nativeFee).toBe(BigInt(maxUint192));
            (0, globals_1.expect)(decoded.linkFee).toBe(BigInt(maxUint192));
        });
        (0, globals_1.it)("should handle minimum int192 values", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const minInt192 = "-3138550867693340381917894711603833208051177722232017256448"; // -(2^191)
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192"], [
                FEED_IDS.V2,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                minInt192, // price
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            (0, globals_1.expect)(decoded.price).toBe(BigInt(minInt192));
        });
        (0, globals_1.it)("should handle reports with very long hex strings", () => {
            // Create a report with maximum valid data
            const reportBlob = createV3ReportBlob(); // V3 has the most fields
            const fullReport = createFullReport(reportBlob);
            // Should handle long but valid hex strings
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V3)).not.toThrow();
        });
        (0, globals_1.it)("should reject reports with null bytes", () => {
            const reportWithNulls = "0x00000000000000000000000000000000";
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(reportWithNulls, FEED_IDS.V2)).toThrow(errors_1.ReportDecodingError);
        });
    });
    (0, globals_1.describe)("error message quality", () => {
        (0, globals_1.it)("should provide specific error for invalid hex", () => {
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)("invalid-hex", FEED_IDS.V2)).toThrow("Report hex string must start with 0x");
        });
        (0, globals_1.it)("should provide specific error for unknown version", () => {
            const unknownVersionFeedId = "0x00996b4aa7e57ca7b68ae1bf45653f56b656fd3aa335ef7fae696b663f1b8472";
            const reportBlob = createV2ReportBlob();
            const fullReport = createFullReport(reportBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, unknownVersionFeedId)).toThrow("Unknown report version: 0x0099");
        });
        (0, globals_1.it)("should provide specific error for V2 decoding failure", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const malformedBlob = abiCoder.encode(["bytes32"], [FEED_IDS.V2]); // Insufficient data
            const fullReport = createFullReport(malformedBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2)).toThrow("Failed to decode V2 report");
        });
        (0, globals_1.it)("should provide specific error for V3 decoding failure", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const malformedBlob = abiCoder.encode(["bytes32"], [FEED_IDS.V3]); // Insufficient data
            const fullReport = createFullReport(malformedBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V3)).toThrow("Failed to decode V3 report");
        });
        (0, globals_1.it)("should provide specific error for V4 decoding failure", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const malformedBlob = abiCoder.encode(["bytes32"], [FEED_IDS.V4]); // Insufficient data
            const fullReport = createFullReport(malformedBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V4)).toThrow("Failed to decode V4 report");
        });
        (0, globals_1.it)("should provide specific error for invalid market status", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint8"], [
                FEED_IDS.V4,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                "2000000000000000000000",
                255,
            ]);
            const fullReport = createFullReport(reportBlob);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V4)).toThrow("Invalid market status: 255");
        });
    });
    (0, globals_1.describe)("performance and memory efficiency", () => {
        (0, globals_1.it)("should decode reports efficiently", () => {
            const reportBlob = createV2ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const start = performance.now();
            // Decode 1000 reports
            for (let i = 0; i < 1000; i++) {
                (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            }
            const end = performance.now();
            const duration = end - start;
            // Should complete in reasonable time,
            // time needs to be big enough to handle ci enviroments.
            (0, globals_1.expect)(duration).toBeLessThan(2000);
        });
        (0, globals_1.it)("should handle large report data efficiently", () => {
            // Create report with maximum values to test memory efficiency
            const abiCoder = new ethers_1.AbiCoder();
            const maxValues = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "int192", "int192"], [
                FEED_IDS.V3,
                4294967295, // max uint32
                4294967295, // max uint32
                "6277101735386680763835789423207666416102355444464034512895", // max uint192
                "6277101735386680763835789423207666416102355444464034512895", // max uint192
                4294967295, // max uint32
                "3138550867693340381917894711603833208051177722232017256447", // max int192
                "3138550867693340381917894711603833208051177722232017256447", // max int192
                "3138550867693340381917894711603833208051177722232017256447", // max int192
            ]);
            const fullReport = createFullReport(maxValues);
            (0, globals_1.expect)(() => (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V3)).not.toThrow();
        });
        (0, globals_1.it)("should fail fast for obviously invalid reports", () => {
            const start = performance.now();
            try {
                (0, decoder_1.decodeReport)("invalid", FEED_IDS.V2);
            }
            catch {
                // Expected to throw
            }
            const end = performance.now();
            const duration = end - start;
            // Should fail very quickly (less than 10ms)
            (0, globals_1.expect)(duration).toBeLessThan(10);
        });
    });
    (0, globals_1.describe)("real-world compatibility", () => {
        (0, globals_1.it)("should handle reports in standard format", () => {
            // Test with standard output format structure
            const reportBlob = createV2ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            // Verify all expected fields are present and correctly typed
            (0, globals_1.expect)(typeof decoded.version).toBe("string");
            (0, globals_1.expect)(typeof decoded.nativeFee).toBe("bigint");
            (0, globals_1.expect)(typeof decoded.linkFee).toBe("bigint");
            (0, globals_1.expect)(typeof decoded.expiresAt).toBe("number");
            (0, globals_1.expect)(typeof decoded.price).toBe("bigint");
        });
        (0, globals_1.it)("should handle reports in alternative format", () => {
            // Test with alternative output format structure
            const reportBlob = createV4ReportBlob();
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V4);
            // Verify V4-specific fields
            (0, globals_1.expect)(decoded.version).toBe("V4");
            (0, globals_1.expect)(decoded.marketStatus).toBeDefined();
            (0, globals_1.expect)(typeof decoded.marketStatus).toBe("number");
        });
        (0, globals_1.it)("should maintain precision for financial data", () => {
            // Test with realistic financial values
            const abiCoder = new ethers_1.AbiCoder();
            const realisticPrice = "2000123456789012345678"; // $2000.123456789012345678 with 18 decimals
            const realisticFee = "1234567890123456"; // 0.001234567890123456 ETH
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192"], [FEED_IDS.V2, 1640995200, 1640995300, realisticFee, realisticFee, 1640995400, realisticPrice]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V2);
            // Verify precision is maintained
            (0, globals_1.expect)(decoded.price.toString()).toBe(realisticPrice);
            (0, globals_1.expect)(decoded.nativeFee.toString()).toBe(realisticFee);
            (0, globals_1.expect)(decoded.linkFee.toString()).toBe(realisticFee);
        });
        (0, globals_1.it)("should handle maximum uint64 values for V8 lastUpdateTimestamp", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const maxUint64 = "18446744073709551615"; // 2^64 - 1
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "uint64", "int192", "uint32"], [
                FEED_IDS.V8,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                maxUint64, // lastUpdateTimestamp
                "2500000000000000000000",
                types_1.MarketStatus.ACTIVE,
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V8);
            (0, globals_1.expect)(decoded.lastUpdateTimestamp).toBe(Number(maxUint64));
        });
        (0, globals_1.it)("should handle large V8 midPrice values", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const largeMidPrice = "3138550867693340381917894711603833208051177722232017256447"; // Large int192
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "uint64", "int192", "uint32"], [
                FEED_IDS.V8,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                1640995250,
                largeMidPrice,
                types_1.MarketStatus.ACTIVE,
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V8);
            (0, globals_1.expect)(decoded.midPrice).toBe(BigInt(largeMidPrice));
        });
        (0, globals_1.it)("should handle large V9 AUM values", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const largeAum = "1000000000000000000000000000000000000000000"; // Very large AUM
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint64", "int192", "uint32"], [
                FEED_IDS.V9,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                "1050000000000000000",
                1640995250,
                largeAum,
                0,
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V9);
            (0, globals_1.expect)(decoded.aum).toBe(BigInt(largeAum));
        });
        (0, globals_1.it)("should handle maximum uint64 values for V9 navDate", () => {
            const abiCoder = new ethers_1.AbiCoder();
            const maxUint64 = "18446744073709551615"; // 2^64 - 1
            const reportBlob = abiCoder.encode(["bytes32", "uint32", "uint32", "uint192", "uint192", "uint32", "int192", "uint64", "int192", "uint32"], [
                FEED_IDS.V9,
                1640995200,
                1640995300,
                "1000000000000000000",
                "500000000000000000",
                1640995400,
                "1050000000000000000",
                maxUint64, // navDate
                "100000000000000000000000",
                0,
            ]);
            const fullReport = createFullReport(reportBlob);
            const decoded = (0, decoder_1.decodeReport)(fullReport, FEED_IDS.V9);
            (0, globals_1.expect)(decoded.navDate).toBe(Number(maxUint64));
        });
    });
});
