// @chainlink/data-streams-sdk/dist/tests/utils/mockWebSocketServer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockWebSocketServer = void 0;
const ws_1 = require("ws");
class MockWebSocketServer {
    constructor() {
        this.wss = null;
        this.clients = new Set();
        this.address = "";
        this.isReady = false;
        this.start();
    }
    start() {
        if (this.wss) {
            return;
        }
        this.wss = new ws_1.WebSocketServer({ port: 0 });
        this.setupServer();
    }
    setupServer() {
        if (!this.wss)
            return;
        this.wss.on("listening", () => {
            if (!this.wss)
                return;
            const addr = this.wss.address();
            this.address = `127.0.0.1:${addr.port}`;
            this.isReady = true;
        });
        this.wss.on("error", error => {
            console.error("WebSocket server error:", error);
            this.isReady = false;
        });
        this.wss.on("connection", (ws) => {
            // Set up error handling first
            ws.on("error", error => {
                console.error("WebSocket connection error:", error);
                this.clients.delete(ws);
                try {
                    ws.close();
                }
                catch {
                    // Ignore close errors
                }
            });
            // Handle pings to keep connection alive
            ws.on("ping", () => {
                try {
                    ws.pong();
                }
                catch {
                    // Ignore pong errors
                }
            });
            this.clients.add(ws);
            ws.on("close", () => {
                this.clients.delete(ws);
            });
        });
    }
    async waitForReady(timeout = 5000) {
        const startTime = Date.now();
        while (Date.now() - startTime < timeout) {
            if (this.isReady && this.address) {
                return true;
            }
            await new Promise(resolve => setTimeout(resolve, 100));
        }
        return false;
    }
    getAddress() {
        return this.isReady ? this.address : "";
    }
    broadcast(data) {
        if (!this.isReady)
            return;
        const deadClients = new Set();
        this.clients.forEach(client => {
            if (client.readyState === ws_1.WebSocket.OPEN) {
                try {
                    client.send(data);
                }
                catch (error) {
                    console.error("Error broadcasting to client:", error);
                    deadClients.add(client);
                }
            }
            else {
                deadClients.add(client);
            }
        });
        // Clean up dead clients
        deadClients.forEach(client => {
            this.clients.delete(client);
            try {
                client.close();
            }
            catch {
                // Ignore close errors
            }
        });
    }
    async closeAllConnections() {
        const closePromises = [];
        this.clients.forEach(client => {
            closePromises.push(new Promise(resolve => {
                const timeout = setTimeout(() => {
                    try {
                        client.terminate();
                    }
                    catch {
                        // Ignore terminate errors
                    }
                    resolve();
                }, 1000);
                client.on("close", () => {
                    clearTimeout(timeout);
                    resolve();
                });
                try {
                    client.close();
                }
                catch {
                    clearTimeout(timeout);
                    resolve();
                }
            }));
        });
        await Promise.all(closePromises);
        this.clients.clear();
    }
    /**
     * Simulate temporary disconnections that trigger reconnection attempts
     * Unlike closeAllConnections, this doesn't clear the client set entirely
     */
    simulateNetworkInterruption() {
        const clientsToDisconnect = Array.from(this.clients);
        clientsToDisconnect.forEach(client => {
            if (client.readyState === ws_1.WebSocket.OPEN) {
                try {
                    // Send close frame to simulate network disconnection
                    client.close(1006, "Network interruption simulation");
                }
                catch {
                    // Ignore close errors
                }
            }
        });
        // Remove disconnected clients from tracking
        // but don't clear the entire set - new connections can still come in
        this.clients.clear();
    }
    /**
     * Simulate realistic connection drops that will trigger SDK reconnection logic
     * This closes connections but keeps the server running to accept reconnections
     */
    simulateConnectionDrops(dropCount) {
        const clientsArray = Array.from(this.clients);
        const toDrop = dropCount || clientsArray.length;
        for (let i = 0; i < Math.min(toDrop, clientsArray.length); i++) {
            const client = clientsArray[i];
            if (client.readyState === ws_1.WebSocket.OPEN) {
                try {
                    // Simulate unexpected disconnection (code 1006 = abnormal closure)
                    client.terminate(); // Force close without handshake
                }
                catch {
                    // Ignore errors
                }
            }
            this.clients.delete(client);
        }
    }
    /**
     * Get current connection count for testing
     */
    getActiveConnectionCount() {
        return Array.from(this.clients).filter(client => client.readyState === ws_1.WebSocket.OPEN).length;
    }
    async close() {
        this.isReady = false;
        this.address = "";
        await this.closeAllConnections();
        return new Promise(resolve => {
            if (!this.wss) {
                resolve();
                return;
            }
            const wss = this.wss;
            this.wss = null;
            const timeout = setTimeout(() => {
                try {
                    wss.close();
                }
                catch {
                    // Ignore close errors
                }
                resolve();
            }, 1000);
            try {
                wss.close(() => {
                    clearTimeout(timeout);
                    resolve();
                });
            }
            catch {
                clearTimeout(timeout);
                resolve();
            }
        });
    }
    getConnectedClientsCount() {
        return Array.from(this.clients).filter(client => client.readyState === ws_1.WebSocket.OPEN).length;
    }
}
exports.MockWebSocketServer = MockWebSocketServer;
