// @chainlink/data-streams-sdk/dist/src/stream/deduplication.js
"use strict";
/**
 * Report deduplication using watermark timestamps
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportDeduplicator = void 0;
/**
 * Manages report deduplication using watermark timestamps
 */
class ReportDeduplicator {
    constructor(options = {}) {
        this.waterMark = new Map();
        this.acceptedCount = 0;
        this.deduplicatedCount = 0;
        this.cleanupInterval = null;
        this.maxWatermarkAge = options.maxWatermarkAge ?? 60 * 60 * 1000; // 1 hour
        this.cleanupIntervalMs = options.cleanupIntervalMs ?? 5 * 60 * 1000; // 5 minutes
        // Start periodic cleanup
        this.startCleanup();
    }
    /**
     * Process a report and determine if it should be accepted or deduplicated
     */
    processReport(report) {
        const feedId = report.feedID;
        const observationsTimestamp = report.observationsTimestamp;
        // Get current watermark for this feed
        const currentWatermark = this.waterMark.get(feedId);
        // Check if this report is older than or equal to the watermark
        if (currentWatermark !== undefined && currentWatermark >= observationsTimestamp) {
            this.deduplicatedCount++;
            return {
                isAccepted: false,
                isDuplicate: true,
                reason: `Report timestamp ${observationsTimestamp} <= watermark ${currentWatermark} for feed ${feedId}`,
            };
        }
        // Accept the report and update watermark
        this.waterMark.set(feedId, observationsTimestamp);
        this.acceptedCount++;
        return {
            isAccepted: true,
            isDuplicate: false,
        };
    }
    /**
     * Get current deduplication statistics
     */
    getStats() {
        return {
            accepted: this.acceptedCount,
            deduplicated: this.deduplicatedCount,
            totalReceived: this.acceptedCount + this.deduplicatedCount,
            watermarkCount: this.waterMark.size,
        };
    }
    /**
     * Get watermark for a specific feed ID
     */
    getWatermark(feedId) {
        return this.waterMark.get(feedId);
    }
    /**
     * Get all current watermarks (for debugging/monitoring)
     */
    getAllWatermarks() {
        const watermarks = {};
        for (const [feedId, timestamp] of this.waterMark) {
            watermarks[feedId] = timestamp;
        }
        return watermarks;
    }
    /**
     * Manually set watermark for a feed (useful for initialization)
     */
    setWatermark(feedId, timestamp) {
        this.waterMark.set(feedId, timestamp);
    }
    /**
     * Clear watermark for a specific feed
     */
    clearWatermark(feedId) {
        return this.waterMark.delete(feedId);
    }
    /**
     * Clear all watermarks
     */
    clearAllWatermarks() {
        this.waterMark.clear();
    }
    /**
     * Reset all counters and watermarks
     */
    reset() {
        this.acceptedCount = 0;
        this.deduplicatedCount = 0;
        this.waterMark.clear();
    }
    /**
     * Start periodic cleanup of old watermarks
     * This prevents memory leaks for feeds that are no longer active
     */
    startCleanup() {
        this.cleanupInterval = setInterval(() => {
            this.cleanupOldWatermarks();
        }, this.cleanupIntervalMs);
    }
    /**
     * Clean up watermarks that are too old
     * This is a safety mechanism to prevent unbounded memory growth
     */
    cleanupOldWatermarks() {
        const now = Date.now();
        const cutoffTime = now - this.maxWatermarkAge;
        // Convert cutoff time to seconds (like the timestamps in reports)
        const cutoffTimestamp = Math.floor(cutoffTime / 1000);
        let _removedCount = 0;
        for (const [feedId, timestamp] of this.waterMark) {
            if (timestamp < cutoffTimestamp) {
                this.waterMark.delete(feedId);
                _removedCount++;
            }
        }
    }
    /**
     * Stop the deduplicator and clean up resources
     */
    stop() {
        if (this.cleanupInterval) {
            clearInterval(this.cleanupInterval);
            this.cleanupInterval = null;
        }
    }
    /**
     * Get memory usage information
     */
    getMemoryInfo() {
        const watermarkCount = this.waterMark.size;
        // Rough estimation: each entry has a string key (~64 chars) + number value
        // String: ~64 bytes (feed ID) + Number: 8 bytes + Map overhead: ~32 bytes
        const estimatedMemoryBytes = watermarkCount * (64 + 8 + 32);
        return {
            watermarkCount,
            estimatedMemoryBytes,
        };
    }
    /**
     * Export watermarks for persistence/debugging
     */
    exportWatermarks() {
        return Array.from(this.waterMark.entries()).map(([feedId, timestamp]) => ({
            feedId,
            timestamp,
        }));
    }
    /**
     * Import watermarks from external source
     */
    importWatermarks(watermarks) {
        for (const { feedId, timestamp } of watermarks) {
            this.waterMark.set(feedId, timestamp);
        }
    }
}
exports.ReportDeduplicator = ReportDeduplicator;
