// @chainlink/data-streams-sdk/dist/src/stream/stats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamStats = void 0;
const metrics_1 = require("../types/metrics");
/**
 * Class for tracking WebSocket connection and report statistics
 */
class StreamStats {
    constructor(configuredConnections = 1) {
        this._accepted = 0;
        this._deduplicated = 0;
        this._partialReconnects = 0;
        this._fullReconnects = 0;
        this._configuredConnections = 0;
        this._activeConnections = 0;
        this._originStatus = {};
        this._totalReceived = 0; // accepted + deduplicated
        this._configuredConnections = configuredConnections;
    }
    /**
     * Increment the number of accepted reports (messages that passed deduplication)
     */
    incrementAccepted() {
        this._accepted++;
        this._totalReceived++;
    }
    /**
     * Increment the number of deduplicated reports (duplicate messages filtered out)
     */
    incrementDeduplicated() {
        this._deduplicated++;
        this._totalReceived++;
    }
    /**
     * Increment the number of partial reconnects (some connections lost but not all)
     */
    incrementPartialReconnects() {
        this._partialReconnects++;
    }
    /**
     * Increment the number of full reconnects (all connections lost)
     */
    incrementFullReconnects() {
        this._fullReconnects++;
    }
    /**
     * Set the number of active connections
     */
    setActiveConnections(count) {
        this._activeConnections = count;
    }
    /**
     * Set the number of configured connections
     */
    setConfiguredConnections(count) {
        this._configuredConnections = count;
    }
    /**
     * Update connection status for a specific origin
     */
    setOriginStatus(origin, status) {
        this._originStatus[origin] = status;
    }
    /**
     * Remove origin status tracking (when origin is no longer used)
     */
    removeOriginStatus(origin) {
        delete this._originStatus[origin];
    }
    /**
     * Get connection status for a specific origin
     */
    getOriginStatus(origin) {
        return this._originStatus[origin] || metrics_1.ConnectionStatus.DISCONNECTED;
    }
    /**
     * Get all origin statuses
     */
    getAllOriginStatuses() {
        return { ...this._originStatus };
    }
    /**
     * Update all origin statuses from ConnectionManager (for unified metrics)
     */
    updateOriginStatuses(originStatus) {
        this._originStatus = { ...originStatus };
    }
    /**
     * Reset all statistics (useful for testing or reconnection scenarios)
     */
    reset() {
        this._accepted = 0;
        this._deduplicated = 0;
        this._partialReconnects = 0;
        this._fullReconnects = 0;
        this._totalReceived = 0;
        this._activeConnections = 0;
        this._originStatus = {};
    }
    /**
     * Get a snapshot of current stream metrics.
     *
     * Returns an immutable snapshot of all metrics
     * The returned object follows the MetricsSnapshot interface for type safety
     *
     * @returns Immutable metrics snapshot
     */
    getStats() {
        return {
            accepted: this._accepted,
            deduplicated: this._deduplicated,
            partialReconnects: this._partialReconnects,
            fullReconnects: this._fullReconnects,
            configuredConnections: this._configuredConnections,
            activeConnections: this._activeConnections,
            totalReceived: this._totalReceived,
            originStatus: Object.freeze({ ...this._originStatus }),
        };
    }
}
exports.StreamStats = StreamStats;
