// @chainlink/data-streams-sdk/dist/tests/unit/stream/connection-manager/auth-handshake.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const connection_manager_1 = require("../../../../src/stream/connection-manager");
const originDiscovery = __importStar(require("../../../../src/utils/origin-discovery"));
// Store event handlers registered via ws.once / ws.on
const onceHandlers = {};
const onHandlers = {};
// Mock WebSocket instance/constructor
const mockWsInstance = {
    on: globals_1.jest.fn((...args) => {
        const [event, cb] = args;
        onHandlers[event] = cb;
    }),
    once: globals_1.jest.fn((...args) => {
        const [event, cb] = args;
        onceHandlers[event] = cb;
    }),
    terminate: globals_1.jest.fn(),
    close: globals_1.jest.fn(),
    get readyState() {
        return 1; // OPEN
    },
};
globals_1.jest.mock("ws", () => {
    const ctor = globals_1.jest.fn(() => mockWsInstance);
    ctor.OPEN = 1;
    ctor.CONNECTING = 0;
    return { __esModule: true, default: ctor };
});
globals_1.jest.mock("../../../../src/utils/origin-discovery", () => ({
    getAvailableOrigins: globals_1.jest.fn(),
}));
(0, globals_1.describe)("ConnectionManager - auth handshake errors", () => {
    let manager;
    let config;
    (0, globals_1.beforeEach)(() => {
        globals_1.jest.clearAllMocks();
        Object.keys(onceHandlers).forEach(k => delete onceHandlers[k]);
        Object.keys(onHandlers).forEach(k => delete onHandlers[k]);
        config = {
            apiKey: "key",
            userSecret: "secret",
            endpoint: "https://api.example.com",
            wsEndpoint: "wss://ws.example.com",
            haMode: false,
        };
        manager = new connection_manager_1.ConnectionManager(config, {
            feedIds: ["0x0003" + "1".repeat(60)],
            maxReconnectAttempts: 2,
            reconnectInterval: 200,
            connectTimeout: 200,
            haMode: false,
            haConnectionTimeout: 200,
        });
        originDiscovery.getAvailableOrigins.mockResolvedValue(["wss://ws.example.com"]);
    });
    (0, globals_1.it)("rejects with WebSocketError on 401 unexpected-response during handshake", async () => {
        const initPromise = manager.initialize();
        // Allow event subscriptions to be registered
        await new Promise(resolve => setTimeout(resolve, 0));
        // Simulate handshake failure via 'unexpected-response'
        const handler = onceHandlers["unexpected-response"];
        (0, globals_1.expect)(typeof handler).toBe("function");
        if (handler) {
            handler({}, { statusCode: 401 });
        }
        await (0, globals_1.expect)(initPromise).rejects.toThrow(/Failed to initialize connections: Failed to establish any WebSocket connections/);
    });
    (0, globals_1.it)("rejects with WebSocketError on 403 unexpected-response during handshake", async () => {
        const initPromise = manager.initialize();
        await new Promise(resolve => setTimeout(resolve, 0));
        // Simulate handshake failure via 'unexpected-response'
        const handler = onceHandlers["unexpected-response"];
        (0, globals_1.expect)(typeof handler).toBe("function");
        if (handler) {
            handler({}, { statusCode: 403 });
        }
        await (0, globals_1.expect)(initPromise).rejects.toThrow(/Failed to initialize connections: Failed to establish any WebSocket connections/);
    });
});
