// @chainlink/data-streams-sdk/dist/tests/unit/stream/connection-manager/health.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const constants_1 = require("../../../../src/utils/constants");
const connection_manager_1 = require("../../../../src/stream/connection-manager");
const originDiscovery = __importStar(require("../../../../src/utils/origin-discovery"));
// Mock WebSocket instance
const mockWsInstance = {
    on: globals_1.jest.fn(),
    once: globals_1.jest.fn(),
    ping: globals_1.jest.fn(),
    pong: globals_1.jest.fn(),
    terminate: globals_1.jest.fn(),
    get readyState() {
        return 1;
    }, // WebSocket.OPEN
};
globals_1.jest.mock("ws", () => {
    const ctor = globals_1.jest.fn(() => mockWsInstance);
    // Provide static constants used by the implementation for readyState checks
    // e.g., WebSocket.OPEN and WebSocket.CONNECTING
    ctor.OPEN = 1;
    ctor.CONNECTING = 0;
    return {
        __esModule: true,
        default: ctor,
    };
});
// Mock origin discovery
globals_1.jest.mock("../../../../src/utils/origin-discovery", () => ({
    getAvailableOrigins: globals_1.jest.fn(),
}));
(0, globals_1.describe)("ConnectionManager - ping/pong health", () => {
    let manager;
    let config;
    (0, globals_1.beforeEach)(() => {
        globals_1.jest.clearAllMocks();
        config = {
            apiKey: "key",
            userSecret: "secret",
            endpoint: "https://api.example.com",
            wsEndpoint: "wss://ws.example.com",
            haMode: false,
        };
        manager = new connection_manager_1.ConnectionManager(config, {
            feedIds: ["0x0003" + "1".repeat(60)],
            maxReconnectAttempts: 3,
            reconnectInterval: 200,
            connectTimeout: 200,
            haMode: false,
            haConnectionTimeout: 200,
        });
        // Mock the origin discovery function to return immediately
        originDiscovery.getAvailableOrigins.mockResolvedValue(["wss://ws.example.com"]);
    });
    (0, globals_1.it)("sets up ping/pong health monitoring after connection", async () => {
        // Mock successful WebSocket connection
        mockWsInstance.on.mockImplementation((...args) => {
            const [event, callback] = args;
            if (event === "open") {
                // Simulate immediate connection
                setTimeout(() => callback(), 1);
            }
        });
        await manager.initialize();
        // Verify that ping and pong event handlers are set up
        (0, globals_1.expect)(mockWsInstance.on).toHaveBeenCalledWith("ping", globals_1.expect.any(Function));
        (0, globals_1.expect)(mockWsInstance.on).toHaveBeenCalledWith("pong", globals_1.expect.any(Function));
    });
    (0, globals_1.it)("responds with pong when server sends ping", async () => {
        const pongSpy = globals_1.jest.fn();
        // Override the pong method to capture calls
        mockWsInstance.pong = pongSpy;
        // Mock successful WebSocket connection; record all subscriptions
        mockWsInstance.on.mockImplementation((...args) => {
            const [event, callback] = args;
            if (event === "open") {
                // Simulate immediate connection
                setTimeout(() => callback(), 1);
            }
        });
        await manager.initialize();
        // Extract the registered ping handler directly from mock.calls
        const pingCall = mockWsInstance.on.mock.calls
            .reverse()
            .find(([evt]) => evt === "ping");
        (0, globals_1.expect)(pingCall).toBeDefined();
        const pingHandler = pingCall && pingCall[1];
        (0, globals_1.expect)(typeof pingHandler).toBe("function");
        // Trigger ping
        if (pingHandler) {
            pingHandler(Buffer.from("data"));
        }
        (0, globals_1.expect)(pongSpy).toHaveBeenCalled();
    });
    (0, globals_1.it)("terminates connection on pong timeout and schedules reconnection", async () => {
        globals_1.jest.useFakeTimers();
        let closeHandler;
        // Mock successful WebSocket connection and capture close handler
        mockWsInstance.on.mockImplementation((...args) => {
            const [event, callback] = args;
            if (event === "open") {
                // With fake timers active, invoke immediately to resolve initialize()
                callback();
            }
            else if (event === "close") {
                closeHandler = callback;
            }
        });
        const reconnectingSpy = globals_1.jest.spyOn(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        manager, "emit");
        await manager.initialize();
        // Advance timers to trigger a ping, then the pong timeout
        globals_1.jest.advanceTimersByTime(constants_1.WS_CONSTANTS.PING_INTERVAL + constants_1.WS_CONSTANTS.PONG_TIMEOUT + 1);
        // Connection should be terminated due to missing pong
        (0, globals_1.expect)(mockWsInstance.terminate).toHaveBeenCalledTimes(1);
        // Simulate the underlying socket closing to trigger reconnection scheduling
        if (closeHandler) {
            closeHandler();
        }
        // Reconnection should be scheduled (reconnecting event emitted with info object)
        const calls = reconnectingSpy.mock.calls;
        const reconnectingCall = calls.find(call => call[0] === "reconnecting");
        (0, globals_1.expect)(reconnectingCall).toBeDefined();
        const info = reconnectingCall && reconnectingCall[1];
        (0, globals_1.expect)(info).toBeDefined();
        if (info) {
            (0, globals_1.expect)(typeof info.attempt).toBe("number");
            (0, globals_1.expect)(typeof info.delayMs).toBe("number");
            (0, globals_1.expect)(typeof info.origin).toBe("string");
            (0, globals_1.expect)(typeof info.host).toBe("string");
        }
        globals_1.jest.useRealTimers();
    });
});
