// @chainlink/data-streams-sdk/dist/tests/unit/stream/connection-manager/max-attempts.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const connection_manager_1 = require("../../../../src/stream/connection-manager");
// No WebSocket connections needed for this test; we exercise internal scheduling logic
(0, globals_1.describe)("ConnectionManager - max reconnect attempts (terminal state)", () => {
    let manager;
    let config;
    (0, globals_1.beforeEach)(() => {
        globals_1.jest.useFakeTimers();
        globals_1.jest.clearAllMocks();
        config = {
            apiKey: "test-key",
            userSecret: "test-secret",
            endpoint: "https://api.example.com",
            wsEndpoint: "wss://ws.example.com",
            haMode: false,
        };
        manager = new connection_manager_1.ConnectionManager(config, {
            feedIds: ["0x0003" + "1".repeat(60)],
            maxReconnectAttempts: 1, // small to hit terminal quickly
            reconnectInterval: 200, // base (doesn't matter, timers are mocked)
            connectTimeout: 500,
            haMode: false,
            haConnectionTimeout: 500,
        });
    });
    (0, globals_1.it)("emits max-reconnect-attempts-reached and all-connections-lost, and sets state to FAILED", () => {
        // Create a fake managed connection and insert into manager internals
        const connection = {
            id: "conn-0",
            origin: "wss://ws.example.com",
            host: "ws.example.com",
            ws: null,
            state: "disconnected",
            reconnectAttempts: 0,
        };
        // Inject into internal map
        manager.connections.set(connection.id, connection);
        // Spy on emit to capture events
        const emitSpy = globals_1.jest.spyOn(manager, "emit");
        // Call private method via any to schedule reconnection
        manager.scheduleReconnection(connection);
        // First schedule increments attempts to 1 and sets a timeout; run timers
        globals_1.jest.runOnlyPendingTimers();
        // Verify terminal events were emitted
        const emittedEvents = emitSpy.mock.calls.map((call) => call[0]);
        (0, globals_1.expect)(emittedEvents).toContain("max-reconnect-attempts-reached");
        (0, globals_1.expect)(emittedEvents).toContain("all-connections-lost");
        // Verify state transitioned to FAILED
        (0, globals_1.expect)(connection.state).toBe("failed");
    });
});
