// @chainlink/data-streams-sdk/dist/tests/unit/stream/connection-manager/partial-failure.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const connection_manager_1 = require("../../../../src/stream/connection-manager");
const ws_1 = __importDefault(require("ws"));
const wsMocks = [];
globals_1.jest.mock("ws", () => ({
    __esModule: true,
    default: globals_1.jest.fn(() => {
        const instance = {
            on: globals_1.jest.fn((_event, _cb) => { }),
            once: globals_1.jest.fn((_event, _cb) => { }),
            terminate: globals_1.jest.fn(),
            close: globals_1.jest.fn(),
            readyState: 1,
        };
        wsMocks.push(instance);
        return instance;
    }),
}));
// Mock origin discovery to return multiple origins
globals_1.jest.mock("../../../../src/utils/origin-discovery", () => {
    return {
        getAvailableOrigins: (..._args) => Promise.resolve(["wss://o1.example.com", "wss://o2.example.com", "wss://o3.example.com"]),
    };
});
(0, globals_1.describe)("ConnectionManager - partial failure emission", () => {
    (0, globals_1.beforeEach)(() => {
        globals_1.jest.clearAllMocks();
        wsMocks.length = 0;
    });
    (0, globals_1.it)("emits partial-failure when some origins fail to establish", async () => {
        const config = {
            apiKey: "k",
            userSecret: "s",
            endpoint: "https://api.example.com",
            wsEndpoint: "wss://o1.example.com,wss://o2.example.com,wss://o3.example.com",
            haMode: true,
        };
        const manager = new connection_manager_1.ConnectionManager(config, {
            feedIds: ["0x0003" + "1".repeat(60)],
            maxReconnectAttempts: 2,
            reconnectInterval: 200,
            connectTimeout: 200,
            haMode: true,
            haConnectionTimeout: 200,
        });
        const emitSpy = globals_1.jest.spyOn(manager, "emit");
        // Configure ws instances so that some fail during establishConnection
        // We simulate failure by triggering 'error' before 'open' for some sockets.
        ws_1.default.mockImplementationOnce(() => {
            const ws = {
                on: globals_1.jest.fn((event, cb) => {
                    if (event === "error")
                        setTimeout(() => cb(new Error("fail o1")), 0);
                }),
                once: globals_1.jest.fn((_event, _cb) => { }),
                terminate: globals_1.jest.fn(),
                close: globals_1.jest.fn(),
                readyState: 1,
            };
            return ws;
        });
        ws_1.default.mockImplementationOnce(() => {
            const ws = {
                on: globals_1.jest.fn((event, cb) => {
                    if (event === "open")
                        setTimeout(() => cb(), 0);
                }),
                once: globals_1.jest.fn((_event, _cb) => { }),
                terminate: globals_1.jest.fn(),
                close: globals_1.jest.fn(),
                readyState: 1,
            };
            return ws;
        });
        ws_1.default.mockImplementationOnce(() => {
            const ws = {
                on: globals_1.jest.fn((event, cb) => {
                    if (event === "error")
                        setTimeout(() => cb(new Error("fail o3")), 0);
                }),
                once: globals_1.jest.fn((_event, _cb) => { }),
                terminate: globals_1.jest.fn(),
                close: globals_1.jest.fn(),
                readyState: 1,
            };
            return ws;
        });
        await manager.initialize().catch(() => { });
        // partial-failure should be emitted with failed and total counts
        const calls = emitSpy.mock.calls.filter((call) => call[0] === "partial-failure");
        (0, globals_1.expect)(calls.length).toBeGreaterThan(0);
        const first = calls[0];
        (0, globals_1.expect)(first[0]).toBe("partial-failure");
        (0, globals_1.expect)(typeof first[1]).toBe("number");
        (0, globals_1.expect)(typeof first[2]).toBe("number");
    });
});
