// @chainlink/data-streams-sdk/dist/tests/unit/stream/connection-manager/status-callback.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const connection_manager_1 = require("../../../../src/stream/connection-manager");
const originDiscovery = __importStar(require("../../../../src/utils/origin-discovery"));
// Mock WebSocket instance
const wsMock = {
    on: globals_1.jest.fn(),
    once: globals_1.jest.fn(),
    readyState: 1,
    ping: globals_1.jest.fn(),
    pong: globals_1.jest.fn(),
    terminate: globals_1.jest.fn(),
};
globals_1.jest.mock("ws", () => ({
    __esModule: true,
    default: globals_1.jest.fn(() => wsMock),
}));
globals_1.jest.mock("../../../../src/utils/origin-discovery", () => ({
    getAvailableOrigins: globals_1.jest.fn(),
}));
(0, globals_1.describe)("ConnectionManager - statusCallback de-duplication", () => {
    let manager;
    let config;
    let statusCallback;
    (0, globals_1.beforeEach)(() => {
        globals_1.jest.clearAllMocks();
        statusCallback = globals_1.jest.fn();
        config = {
            apiKey: "key",
            userSecret: "secret",
            endpoint: "https://api.example.com",
            wsEndpoint: "wss://ws.example.com",
            haMode: false,
            connectionStatusCallback: statusCallback,
        };
        manager = new connection_manager_1.ConnectionManager(config, {
            feedIds: ["0x0003" + "1".repeat(60)],
            maxReconnectAttempts: 2,
            reconnectInterval: 200,
            connectTimeout: 200,
            haMode: false,
            haConnectionTimeout: 200,
            statusCallback,
        });
        // Mock the origin discovery function to return immediately
        originDiscovery.getAvailableOrigins.mockResolvedValue(["wss://ws.example.com"]);
    });
    (0, globals_1.it)("invokes callback once for connect and once for disconnect", async () => {
        let closeHandler;
        // Mock successful WebSocket connection and capture close handler
        wsMock.on.mockImplementation((...args) => {
            const [event, callback] = args;
            if (event === "open") {
                // Simulate immediate connection
                setTimeout(() => callback(), 1);
            }
            else if (event === "close") {
                closeHandler = callback;
            }
        });
        await manager.initialize();
        (0, globals_1.expect)(statusCallback).toHaveBeenCalledTimes(1);
        (0, globals_1.expect)(statusCallback).toHaveBeenLastCalledWith(true, globals_1.expect.any(String), globals_1.expect.any(String));
        // Trigger close event
        if (closeHandler) {
            closeHandler();
        }
        // After close, callback should be called once more with false
        (0, globals_1.expect)(statusCallback).toHaveBeenCalledTimes(2);
        (0, globals_1.expect)(statusCallback).toHaveBeenLastCalledWith(false, globals_1.expect.any(String), globals_1.expect.any(String));
    });
});
