// @chainlink/data-streams-sdk/dist/tests/unit/stream/stream/events.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const stream_1 = require("../../../../src/stream");
const logger_1 = require("../../../../src/types/logger");
const originDiscovery = __importStar(require("../../../../src/utils/origin-discovery"));
// Mock ConnectionManager to control event emission
const mockConnectionManager = {
    on: globals_1.jest.fn(),
    removeAllListeners: globals_1.jest.fn(),
    initialize: globals_1.jest.fn(() => Promise.resolve()),
    shutdown: globals_1.jest.fn(() => Promise.resolve()),
    getActiveConnectionCount: globals_1.jest.fn().mockReturnValue(1),
    getConfiguredConnectionCount: globals_1.jest.fn().mockReturnValue(1),
    getConnectionDetails: globals_1.jest.fn().mockReturnValue([{ origin: "wss://ws.example.com", host: "ws.example.com" }]),
    getOriginStatusMap: globals_1.jest.fn().mockReturnValue({}),
    setStreamStats: globals_1.jest.fn(),
};
// Store event handlers from ConnectionManager
const connectionManagerHandlers = {};
globals_1.jest.mock("../../../../src/stream/connection-manager", () => ({
    ConnectionManager: globals_1.jest.fn(() => {
        mockConnectionManager.on.mockImplementation((...args) => {
            const [event, handler] = args;
            if (!connectionManagerHandlers[event]) {
                connectionManagerHandlers[event] = [];
            }
            connectionManagerHandlers[event].push(handler);
        });
        return mockConnectionManager;
    }),
}));
globals_1.jest.mock("../../../../src/utils/origin-discovery", () => ({
    getAvailableOrigins: globals_1.jest.fn(),
}));
(0, globals_1.describe)("Stream - Event Re-emission", () => {
    let stream;
    let config;
    (0, globals_1.beforeEach)(() => {
        globals_1.jest.clearAllMocks();
        Object.keys(connectionManagerHandlers).forEach(key => delete connectionManagerHandlers[key]);
        const silent = { debug: () => { }, info: () => { }, warn: () => { }, error: () => { } };
        config = {
            apiKey: "test-key",
            userSecret: "test-secret",
            endpoint: "https://api.example.com",
            wsEndpoint: "wss://ws.example.com",
            logging: {
                logger: silent,
                logLevel: logger_1.LogLevel.ERROR, // Suppress logs in tests
            },
        };
        originDiscovery.getAvailableOrigins.mockResolvedValue(["wss://ws.example.com"]);
        stream = new stream_1.Stream(config, ["0x0003" + "1".repeat(60)]);
    });
    afterEach(async () => {
        try {
            await stream.close();
        }
        catch {
            // Ignore cleanup errors
        }
    });
    (0, globals_1.it)("re-emits 'reconnecting' event exactly once per transition", async () => {
        const reconnectingSpy = globals_1.jest.fn();
        stream.on("reconnecting", reconnectingSpy);
        await stream.connect();
        // Simulate reconnecting event from ConnectionManager
        const reconnectingInfo = { attempt: 1, delayMs: 1000, origin: "wss://ws.example.com", host: "ws.example.com" };
        const reconnectingHandlers = connectionManagerHandlers["reconnecting"] || [];
        (0, globals_1.expect)(reconnectingHandlers).toHaveLength(1);
        // Emit reconnecting event multiple times (should only re-emit once each)
        reconnectingHandlers[0](reconnectingInfo);
        reconnectingHandlers[0](reconnectingInfo);
        (0, globals_1.expect)(reconnectingSpy).toHaveBeenCalledTimes(2);
        (0, globals_1.expect)(reconnectingSpy).toHaveBeenCalledWith(reconnectingInfo);
    });
    (0, globals_1.it)("re-emits 'connection-lost' event exactly once per transition", async () => {
        const connectionLostSpy = globals_1.jest.fn();
        stream.on("connection-lost", connectionLostSpy);
        await stream.connect();
        // Simulate connection-lost event from ConnectionManager
        const mockConnection = { id: "conn-1", origin: "wss://ws.example.com", host: "ws.example.com" };
        const mockError = new Error("Connection lost");
        const connectionLostHandlers = connectionManagerHandlers["connection-lost"] || [];
        (0, globals_1.expect)(connectionLostHandlers).toHaveLength(1);
        connectionLostHandlers[0](mockConnection, mockError);
        (0, globals_1.expect)(connectionLostSpy).toHaveBeenCalledTimes(1);
        (0, globals_1.expect)(connectionLostSpy).toHaveBeenCalledWith(mockConnection, mockError);
    });
    (0, globals_1.it)("re-emits 'all-connections-lost' and 'disconnected' events exactly once per transition", async () => {
        const allConnectionsLostSpy = globals_1.jest.fn();
        const disconnectedSpy = globals_1.jest.fn();
        stream.on("all-connections-lost", allConnectionsLostSpy);
        stream.on("disconnected", disconnectedSpy);
        await stream.connect();
        // Simulate all-connections-lost event from ConnectionManager
        const allConnectionsLostHandlers = connectionManagerHandlers["all-connections-lost"] || [];
        (0, globals_1.expect)(allConnectionsLostHandlers).toHaveLength(1);
        allConnectionsLostHandlers[0]();
        (0, globals_1.expect)(allConnectionsLostSpy).toHaveBeenCalledTimes(1);
        (0, globals_1.expect)(disconnectedSpy).toHaveBeenCalledTimes(1);
    });
    (0, globals_1.it)("handles 'connection-restored' event without re-emission (internal only)", async () => {
        // connection-restored is handled internally for stats but not re-emitted to public API
        const connectionRestoredSpy = globals_1.jest.fn();
        stream.on("connection-restored", connectionRestoredSpy);
        await stream.connect();
        // The Stream class doesn't re-emit connection-restored, so we shouldn't have any public listeners
        (0, globals_1.expect)(connectionRestoredSpy).not.toHaveBeenCalled();
    });
});
