// @chainlink/data-streams-sdk/dist/tests/unit/stream/stream/read.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const stream_1 = require("../../../../src/stream");
const logger_1 = require("../../../../src/types/logger");
const originDiscovery = __importStar(require("../../../../src/utils/origin-discovery"));
// Mock ConnectionManager
const mockConnectionManager = {
    on: globals_1.jest.fn(),
    removeAllListeners: globals_1.jest.fn(),
    initialize: globals_1.jest.fn(() => Promise.resolve()),
    shutdown: globals_1.jest.fn(() => Promise.resolve()),
    getActiveConnectionCount: globals_1.jest.fn().mockReturnValue(1),
    getConfiguredConnectionCount: globals_1.jest.fn().mockReturnValue(1),
    getConnectionDetails: globals_1.jest.fn().mockReturnValue([{ origin: "wss://ws.example.com", host: "ws.example.com" }]),
    getOriginStatusMap: globals_1.jest.fn().mockReturnValue({}),
    setStreamStats: globals_1.jest.fn(),
};
globals_1.jest.mock("../../../../src/stream/connection-manager", () => ({
    ConnectionManager: globals_1.jest.fn(() => mockConnectionManager),
}));
globals_1.jest.mock("../../../../src/utils/origin-discovery", () => ({
    getAvailableOrigins: globals_1.jest.fn(),
}));
(0, globals_1.describe)("Stream - read() method", () => {
    let stream;
    let config;
    (0, globals_1.beforeEach)(async () => {
        globals_1.jest.clearAllMocks();
        // Close any existing stream to ensure clean state
        if (stream) {
            try {
                await stream.close();
            }
            catch {
                // Ignore close errors
            }
        }
        const silent = { debug: () => { }, info: () => { }, warn: () => { }, error: () => { } };
        config = {
            apiKey: "test-key",
            userSecret: "test-secret",
            endpoint: "https://api.example.com",
            wsEndpoint: "wss://ws.example.com",
            logging: {
                logger: silent,
                logLevel: logger_1.LogLevel.ERROR, // Suppress logs in tests
            },
        };
        originDiscovery.getAvailableOrigins.mockResolvedValue(["wss://ws.example.com"]);
        stream = new stream_1.Stream(config, ["0x0003" + "1".repeat(60)]);
    });
    afterEach(async () => {
        try {
            await stream.close();
        }
        catch {
            // Ignore cleanup errors
        }
    });
    (0, globals_1.it)("resolves on next 'report' event and cleans up listeners", async () => {
        await stream.connect();
        const mockReport = {
            feedID: "0x0003" + "1".repeat(60),
            fullReport: "0x" + "a".repeat(64),
            validFromTimestamp: 1234567890,
            observationsTimestamp: 1234567890,
        };
        // Start read() operation
        const readPromise = stream.read();
        // Simulate receiving a report
        setTimeout(() => {
            stream.emit("report", mockReport);
        }, 10);
        // Should resolve with the report
        const result = await readPromise;
        (0, globals_1.expect)(result).toEqual(mockReport);
        // Verify listeners are cleaned up
        (0, globals_1.expect)(stream.listenerCount("report")).toBe(0);
        (0, globals_1.expect)(stream.listenerCount("error")).toBe(0);
    });
    (0, globals_1.it)("rejects on 'error' event and cleans up listeners", async () => {
        await stream.connect();
        const mockError = new Error("Test error");
        // Start read() operation
        const readPromise = stream.read();
        // Simulate an error
        setTimeout(() => {
            stream.emit("error", mockError);
        }, 10);
        // Should reject with the error
        await (0, globals_1.expect)(readPromise).rejects.toThrow("Test error");
        // Verify listeners are cleaned up
        (0, globals_1.expect)(stream.listenerCount("report")).toBe(0);
        (0, globals_1.expect)(stream.listenerCount("error")).toBe(0);
    });
    (0, globals_1.it)("cleans up listeners when report arrives first", async () => {
        await stream.connect();
        const mockReport = {
            feedID: "0x0003" + "1".repeat(60),
            fullReport: "0x" + "b".repeat(64),
            validFromTimestamp: 1234567891,
            observationsTimestamp: 1234567891,
        };
        // Start read() operation
        const readPromise = stream.read();
        // Emit report first - this should resolve immediately and clean up listeners
        stream.emit("report", mockReport);
        // Should resolve with the report
        const result = await readPromise;
        (0, globals_1.expect)(result).toEqual(mockReport);
        // Verify listeners are cleaned up
        (0, globals_1.expect)(stream.listenerCount("report")).toBe(0);
        (0, globals_1.expect)(stream.listenerCount("error")).toBe(0);
    });
    (0, globals_1.it)("cleans up listeners when error arrives first", async () => {
        await stream.connect();
        const mockError = new Error("First error specific test");
        const mockReport = {
            feedID: "0x0003" + "1".repeat(60),
            fullReport: "0x" + "c".repeat(64),
            validFromTimestamp: 1234567892,
            observationsTimestamp: 1234567892,
        };
        // Start read() operation
        const readPromise = stream.read();
        // Emit error first, then report
        setTimeout(() => {
            stream.emit("error", mockError);
            setTimeout(() => {
                stream.emit("report", mockReport);
            }, 0);
        }, 10);
        // Should reject with the error (report ignored after cleanup)
        await (0, globals_1.expect)(readPromise).rejects.toThrow("First error specific test");
        // Verify listeners are cleaned up
        (0, globals_1.expect)(stream.listenerCount("report")).toBe(0);
        (0, globals_1.expect)(stream.listenerCount("error")).toBe(0);
    });
    (0, globals_1.it)("supports multiple concurrent read() operations", async () => {
        await stream.connect();
        const mockReport = {
            feedID: "0x0003" + "1".repeat(60),
            fullReport: "0x" + "d".repeat(64),
            validFromTimestamp: 1234567895,
            observationsTimestamp: 1234567895,
        };
        // Start two concurrent read() operations
        const readPromise1 = stream.read();
        const readPromise2 = stream.read();
        // Should have 2 report listeners and 2 error listeners
        (0, globals_1.expect)(stream.listenerCount("report")).toBe(2);
        (0, globals_1.expect)(stream.listenerCount("error")).toBe(2);
        // Emit one report - both read() operations should resolve with the same report
        stream.emit("report", mockReport);
        stream.emit("report", mockReport); // Emit twice to satisfy both listeners
        // Both should resolve with the same report
        const results = await Promise.all([readPromise1, readPromise2]);
        // Both operations should succeed
        (0, globals_1.expect)(results).toHaveLength(2);
        (0, globals_1.expect)(results[0]).toEqual(mockReport);
        (0, globals_1.expect)(results[1]).toEqual(mockReport);
        // All listeners should be cleaned up
        (0, globals_1.expect)(stream.listenerCount("report")).toBe(0);
        (0, globals_1.expect)(stream.listenerCount("error")).toBe(0);
    });
});
