// @chainlink/data-streams-sdk/dist/tests/unit/stream/stream/terminal.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const stream_1 = require("../../../../src/stream");
const logger_1 = require("../../../../src/types/logger");
const originDiscovery = __importStar(require("../../../../src/utils/origin-discovery"));
// Mock ConnectionManager to control behavior
const mockConnectionManager = {
    on: globals_1.jest.fn(),
    removeAllListeners: globals_1.jest.fn(),
    initialize: globals_1.jest.fn(() => Promise.resolve()),
    shutdown: globals_1.jest.fn(() => Promise.resolve()),
    getActiveConnectionCount: globals_1.jest.fn().mockReturnValue(0), // No active connections
    getConfiguredConnectionCount: globals_1.jest.fn().mockReturnValue(1),
    getConnectionDetails: globals_1.jest.fn().mockReturnValue([{ origin: "wss://ws.example.com", host: "ws.example.com" }]),
    getOriginStatusMap: globals_1.jest.fn().mockReturnValue({}),
    setStreamStats: globals_1.jest.fn(),
};
// Store event handlers from ConnectionManager
const connectionManagerHandlers = {};
globals_1.jest.mock("../../../../src/stream/connection-manager", () => ({
    ConnectionManager: globals_1.jest.fn(() => {
        mockConnectionManager.on.mockImplementation((...args) => {
            const [event, handler] = args;
            if (!connectionManagerHandlers[event]) {
                connectionManagerHandlers[event] = [];
            }
            connectionManagerHandlers[event].push(handler);
        });
        return mockConnectionManager;
    }),
}));
globals_1.jest.mock("../../../../src/utils/origin-discovery", () => ({
    getAvailableOrigins: globals_1.jest.fn(),
}));
(0, globals_1.describe)("Stream - Terminal Disconnected State", () => {
    let stream;
    let config;
    (0, globals_1.beforeEach)(() => {
        globals_1.jest.clearAllMocks();
        Object.keys(connectionManagerHandlers).forEach(key => delete connectionManagerHandlers[key]);
        // Reset the initialize method to successful resolution for each test
        mockConnectionManager.initialize = globals_1.jest.fn(() => Promise.resolve());
        const silent = { debug: () => { }, info: () => { }, warn: () => { }, error: () => { } };
        config = {
            apiKey: "test-key",
            userSecret: "test-secret",
            endpoint: "https://api.example.com",
            wsEndpoint: "wss://ws.example.com",
            logging: {
                logger: silent,
                logLevel: logger_1.LogLevel.ERROR, // Suppress logs in tests
            },
        };
        originDiscovery.getAvailableOrigins.mockResolvedValue(["wss://ws.example.com"]);
        stream = new stream_1.Stream(config, ["0x0003" + "1".repeat(60)]);
    });
    afterEach(async () => {
        try {
            await stream.close();
        }
        catch {
            // Ignore cleanup errors
        }
    });
    (0, globals_1.it)("emits 'disconnected' when all connections are lost terminally", async () => {
        const disconnectedSpy = globals_1.jest.fn();
        const allConnectionsLostSpy = globals_1.jest.fn();
        stream.on("disconnected", disconnectedSpy);
        stream.on("all-connections-lost", allConnectionsLostSpy);
        await stream.connect();
        // Simulate terminal disconnection (all connections lost after max attempts)
        const allConnectionsLostHandlers = connectionManagerHandlers["all-connections-lost"] || [];
        (0, globals_1.expect)(allConnectionsLostHandlers).toHaveLength(1);
        allConnectionsLostHandlers[0]();
        (0, globals_1.expect)(disconnectedSpy).toHaveBeenCalledTimes(1);
        (0, globals_1.expect)(allConnectionsLostSpy).toHaveBeenCalledTimes(1);
    });
    (0, globals_1.it)("does not attempt further reconnections after disconnected state", async () => {
        const reconnectingSpy = globals_1.jest.fn();
        const disconnectedSpy = globals_1.jest.fn();
        stream.on("reconnecting", reconnectingSpy);
        stream.on("disconnected", disconnectedSpy);
        await stream.connect();
        // Simulate terminal disconnection
        const allConnectionsLostHandlers = connectionManagerHandlers["all-connections-lost"] || [];
        allConnectionsLostHandlers[0]();
        (0, globals_1.expect)(disconnectedSpy).toHaveBeenCalledTimes(1);
        // The test concept here is that in practice, the ConnectionManager wouldn't emit
        // more 'reconnecting' events after reaching terminal state, but we can't easily test that
        // without the real ConnectionManager logic. This test passes as the behavior
        // of preventing reconnections is handled by the ConnectionManager itself.
        (0, globals_1.expect)(reconnectingSpy).toHaveBeenCalledTimes(0);
    });
    (0, globals_1.it)("calling connect() again on same instance should not revive connection", async () => {
        const disconnectedSpy = globals_1.jest.fn();
        stream.on("disconnected", disconnectedSpy);
        // Initial connection
        await stream.connect();
        // Simulate terminal disconnection
        const allConnectionsLostHandlers = connectionManagerHandlers["all-connections-lost"] || [];
        allConnectionsLostHandlers[0]();
        (0, globals_1.expect)(disconnectedSpy).toHaveBeenCalledTimes(1);
        // Try to connect again - this should fail or not restart connections
        // The ConnectionManager should be in a terminal state
        mockConnectionManager.initialize = globals_1.jest.fn(() => Promise.reject(new Error("Connection manager is in terminal state")));
        await (0, globals_1.expect)(stream.connect()).rejects.toThrow("Connection manager is in terminal state");
        // Should not emit new connection events
        (0, globals_1.expect)(disconnectedSpy).toHaveBeenCalledTimes(1); // Still only the original disconnection
    });
    (0, globals_1.it)("read() should reject after disconnected state", async () => {
        await stream.connect();
        // Simulate terminal disconnection
        const allConnectionsLostHandlers = connectionManagerHandlers["all-connections-lost"] || [];
        allConnectionsLostHandlers[0]();
        // read() after disconnection should fail gracefully
        const readPromise = stream.read();
        // Simulate that no reports will come
        setTimeout(() => {
            stream.emit("error", new Error("Stream is disconnected"));
        }, 10);
        await (0, globals_1.expect)(readPromise).rejects.toThrow("Stream is disconnected");
    });
    (0, globals_1.it)("getMetrics() should reflect disconnected state", async () => {
        await stream.connect();
        // Check initial state
        let metrics = stream.getMetrics();
        (0, globals_1.expect)(metrics.activeConnections).toBe(0); // Mocked to return 0
        // Simulate terminal disconnection
        const allConnectionsLostHandlers = connectionManagerHandlers["all-connections-lost"] || [];
        allConnectionsLostHandlers[0]();
        // Metrics should still reflect the disconnected state
        metrics = stream.getMetrics();
        (0, globals_1.expect)(metrics.activeConnections).toBe(0);
        (0, globals_1.expect)(metrics.configuredConnections).toBe(1);
    });
    (0, globals_1.it)("close() should work gracefully even after disconnected state", async () => {
        await stream.connect();
        // Simulate terminal disconnection
        const allConnectionsLostHandlers = connectionManagerHandlers["all-connections-lost"] || [];
        allConnectionsLostHandlers[0]();
        // close() should still work
        await (0, globals_1.expect)(stream.close()).resolves.not.toThrow();
        // Verify shutdown was called
        (0, globals_1.expect)(mockConnectionManager.shutdown).toHaveBeenCalled();
    });
});
