// @chainlink/data-streams-sdk/dist/src/client/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseClient = void 0;
const constants_1 = require("../utils/constants");
const errors_1 = require("../types/errors");
const logger_1 = require("../types/logger");
const auth_1 = require("../utils/auth");
const logger_2 = require("../utils/logger");
/**
 * Base client class with common functionality
 */
class BaseClient {
    constructor(config) {
        this.config = config;
        this.logger = new logger_2.SDKLogger(config.logging);
        this.validateConfig(config);
    }
    /**
     * Validate the configuration object
     * @param config Configuration to validate
     * @throws {ValidationError} If configuration is invalid
     */
    validateConfig(config) {
        this.logger.debug("Starting configuration validation");
        // Check if config exists
        if (!config || typeof config !== "object") {
            throw new errors_1.ValidationError("Configuration object is required");
        }
        // Validate required string fields
        this.validateRequiredString(config.apiKey, "apiKey");
        this.validateRequiredString(config.userSecret, "userSecret");
        this.validateRequiredString(config.endpoint, "endpoint");
        this.validateRequiredString(config.wsEndpoint, "wsEndpoint");
        // Validate URLs
        this.validateUrl(config.endpoint, "endpoint", ["http:", "https:"]);
        this.validateWebSocketUrls(config.wsEndpoint);
        // Validate optional numeric fields
        if (config.timeout !== undefined) {
            this.validatePositiveNumber(config.timeout, "timeout");
        }
        if (config.retryAttempts !== undefined) {
            this.validateNonNegativeNumber(config.retryAttempts, "retryAttempts");
        }
        if (config.retryDelay !== undefined) {
            this.validateNonNegativeNumber(config.retryDelay, "retryDelay");
        }
        if (config.haConnectionTimeout !== undefined) {
            this.validatePositiveNumber(config.haConnectionTimeout, "haConnectionTimeout");
        }
        // Validate boolean fields
        if (config.haMode !== undefined && typeof config.haMode !== "boolean") {
            throw new errors_1.ValidationError("haMode must be a boolean");
        }
        // Validate callback function
        if (config.connectionStatusCallback !== undefined && typeof config.connectionStatusCallback !== "function") {
            throw new errors_1.ValidationError("connectionStatusCallback must be a function");
        }
        // Validate logging config
        this.validateLoggingConfig(config.logging);
        this.logger.debug("Configuration validation completed successfully");
    }
    /**
     * Validate a required string field
     */
    validateRequiredString(value, fieldName) {
        if (value === undefined || value === null) {
            throw new errors_1.ValidationError(`${fieldName} is required`);
        }
        if (typeof value !== "string") {
            throw new errors_1.ValidationError(`${fieldName} must be a string`);
        }
        if (value.trim() === "") {
            throw new errors_1.ValidationError(`${fieldName} cannot be empty`);
        }
    }
    /**
     * Validate a URL field
     */
    validateUrl(value, fieldName, allowedProtocols) {
        try {
            const url = new URL(value);
            if (!allowedProtocols.includes(url.protocol)) {
                throw new errors_1.ValidationError(`${fieldName} must use one of these protocols: ${allowedProtocols.join(", ")}`);
            }
        }
        catch (error) {
            if (error instanceof errors_1.ValidationError) {
                throw error;
            }
            throw new errors_1.ValidationError(`${fieldName} must be a valid URL`);
        }
    }
    /**
     * Validate WebSocket URLs
     */
    validateWebSocketUrls(wsEndpoint) {
        const urls = wsEndpoint
            .split(",")
            .map(url => url.trim())
            .filter(url => url.length > 0);
        if (urls.length === 0) {
            throw new errors_1.ValidationError("wsEndpoint cannot be empty");
        }
        for (const url of urls) {
            this.validateUrl(url, "wsEndpoint", ["ws:", "wss:"]);
        }
    }
    /**
     * Validate a positive number
     */
    validatePositiveNumber(value, fieldName) {
        if (typeof value !== "number" || isNaN(value)) {
            throw new errors_1.ValidationError(`${fieldName} must be a number`);
        }
        if (value <= 0) {
            throw new errors_1.ValidationError(`${fieldName} must be positive`);
        }
    }
    /**
     * Validate a non-negative number
     */
    validateNonNegativeNumber(value, fieldName) {
        if (typeof value !== "number" || isNaN(value)) {
            throw new errors_1.ValidationError(`${fieldName} must be a number`);
        }
        if (value < 0) {
            throw new errors_1.ValidationError(`${fieldName} cannot be negative`);
        }
    }
    /**
     * Validate logging configuration
     */
    validateLoggingConfig(logging) {
        if (!logging) {
            return;
        }
        // Validate logLevel
        if (logging.logLevel !== undefined) {
            if (!Object.values(logger_1.LogLevel).includes(logging.logLevel)) {
                throw new errors_1.ValidationError(`Invalid logLevel: ${logging.logLevel}. Must be one of: ${Object.values(logger_1.LogLevel).join(", ")}`);
            }
        }
        // Validate logger functions
        if (logging.logger) {
            const logger = logging.logger;
            const validLevels = ["debug", "info", "warn", "error"];
            validLevels.forEach(level => {
                if (logger[level] && typeof logger[level] !== "function") {
                    throw new errors_1.ValidationError(`Logger.${level} must be a function, got ${typeof logger[level]}`);
                }
            });
        }
        // Validate enableConnectionDebug
        if (logging.enableConnectionDebug !== undefined && typeof logging.enableConnectionDebug !== "boolean") {
            throw new errors_1.ValidationError("enableConnectionDebug must be a boolean");
        }
    }
    /**
     * Make an authenticated HTTP request
     * @param path The API path
     * @param options Request options
     * @returns The response data
     * @throws {APIError} If the request fails
     */
    async makeRequest(path, options = {}) {
        const url = new URL(path, this.config.endpoint);
        const method = options.method || "GET";
        const body = typeof options.body === "string" ? options.body : undefined;
        this.logger.debug(`Making ${method} request to ${url}`);
        // Generate auth headers
        const authHeaders = (0, auth_1.generateAuthHeaders)(this.config.apiKey, this.config.userSecret, method, url.toString(), body);
        const headers = new Headers(options.headers);
        headers.set("Content-Type", "application/json");
        Object.entries(authHeaders).forEach(([key, value]) => {
            headers.set(key, value);
        });
        const response = await fetch(url.toString(), {
            ...options,
            headers,
            signal: AbortSignal.timeout(this.config.timeout ?? constants_1.DEFAULT_TIMEOUT),
        });
        if (!response.ok) {
            let errorMessage;
            try {
                const errorData = await response.json();
                errorMessage = errorData.message || response.statusText;
            }
            catch {
                errorMessage = response.statusText;
            }
            this.logger.error(`Request failed: ${method} ${url} - ${response.status} ${errorMessage}`);
            throw new errors_1.APIError(errorMessage, response.status);
        }
        this.logger.info(`Request successful: ${method} ${url} - ${response.status}`);
        return response.json();
    }
    /**
     * Retry a function with exponential backoff
     * @param fn The function to retry
     * @returns The function result
     * @throws The last error encountered
     */
    async withRetry(fn) {
        const maxAttempts = this.config.retryAttempts ?? constants_1.DEFAULT_RETRY_ATTEMPTS;
        const baseDelay = this.config.retryDelay ?? constants_1.DEFAULT_RETRY_DELAY;
        this.logger.debug(`Starting retry logic: max ${maxAttempts} attempts`);
        let lastError;
        for (let attempt = 0; attempt <= maxAttempts; attempt++) {
            try {
                return await fn();
            }
            catch (error) {
                lastError = error;
                if (attempt === maxAttempts) {
                    this.logger.error(`All retry attempts failed (${maxAttempts}/${maxAttempts}):`, lastError);
                    break;
                }
                // Skip retry for client errors (validation/authentication)
                if (error instanceof errors_1.APIError && (error.statusCode === 400 || error.statusCode === 401)) {
                    this.logger.warn(`Not retrying client error ${error.statusCode}: ${error.message}`);
                    throw error;
                }
                // Exponential backoff with jitter
                const delay = baseDelay * Math.pow(2, attempt) * (0.5 + Math.random() * 0.5);
                this.logger.warn(`Retry attempt ${attempt}/${maxAttempts} failed, retrying in ${Math.round(delay)}ms:`, error);
                await new Promise(resolve => setTimeout(resolve, delay));
            }
        }
        throw lastError;
    }
}
exports.BaseClient = BaseClient;
