// @chainlink/data-streams-sdk/dist/src/client/implementation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataStreamsClientImpl = void 0;
const base_1 = require("./base");
const validation_1 = require("../utils/validation");
const stream_1 = require("../stream");
/**
 * Main implementation of the Data Streams client
 */
class DataStreamsClientImpl extends base_1.BaseClient {
    constructor(config) {
        super(config);
        this.validateHAConfiguration(config);
        this.logger.info("Data Streams client initialized");
    }
    /**
     * Validate HA mode configuration
     */
    validateHAConfiguration(config) {
        if (config.haMode) {
            const origins = this.parseOrigins(config.wsEndpoint);
            if (origins.length === 0) {
                throw new Error("HA mode enabled but no WebSocket endpoints provided");
            }
        }
        // Validate comma-separated URLs format
        if (config.wsEndpoint) {
            const origins = this.parseOrigins(config.wsEndpoint);
            for (const origin of origins) {
                if (!origin.startsWith("ws://") && !origin.startsWith("wss://")) {
                    throw new Error(`Invalid WebSocket URL format: ${origin}. Must start with ws:// or wss://`);
                }
            }
        }
    }
    /**
     * Parse comma-separated WebSocket URLs
     */
    parseOrigins(wsEndpoint) {
        if (!wsEndpoint) {
            return [];
        }
        return wsEndpoint
            .split(",")
            .map(url => url.trim())
            .filter(url => url.length > 0);
    }
    /**
     * List all available feeds
     * @returns Array of available feeds
     */
    async listFeeds() {
        this.logger.debug("Fetching available feeds");
        const response = await this.withRetry(() => this.makeRequest("/api/v1/feeds"));
        this.logger.debug(`Retrieved ${response.feeds.length} feeds`);
        return response.feeds;
    }
    /**
     * Get the latest report for a feed
     * @param feedId The feed ID to get the report for
     * @returns The raw report data
     */
    async getLatestReport(feedId) {
        (0, validation_1.validateFeedId)(feedId);
        this.logger.debug(`Fetching latest report for feed ${feedId}`);
        const response = await this.withRetry(() => this.makeRequest(`/api/v1/reports/latest?feedID=${feedId}`));
        this.logger.debug(`Retrieved latest report for feed ${feedId} (timestamp: ${response.report.observationsTimestamp})`);
        return response.report;
    }
    /**
     * Get a report for a feed at a specific timestamp
     * @param feedId The feed ID to get the report for
     * @param timestamp The timestamp to get the report for
     * @returns The raw report data
     */
    async getReportByTimestamp(feedId, timestamp) {
        (0, validation_1.validateFeedId)(feedId);
        (0, validation_1.validateTimestamp)(timestamp);
        this.logger.debug(`Fetching report for feed ${feedId} at timestamp ${timestamp}`);
        const response = await this.withRetry(() => this.makeRequest(`/api/v1/reports?feedID=${feedId}&timestamp=${timestamp}`));
        this.logger.debug(`Retrieved report for feed ${feedId} at timestamp ${timestamp}`);
        return response.report;
    }
    /**
     * Get a range of reports for a feed
     * @param feedId The feed ID to get reports for
     * @param startTime The start timestamp, inclusive
     * @param limit Maximum number of reports to return
     * @returns Array of raw report data
     */
    async getReportsPage(feedId, startTime, limit = 10) {
        (0, validation_1.validateFeedId)(feedId);
        (0, validation_1.validateTimestamp)(startTime);
        this.logger.debug(`Fetching ${limit} reports for feed ${feedId} starting from timestamp ${startTime}`);
        const response = await this.withRetry(() => this.makeRequest(`/api/v1/reports/page?feedID=${feedId}&startTimestamp=${startTime}&limit=${limit}`));
        this.logger.info(`Retrieved ${response.reports.length} reports for feed ${feedId} (requested: ${limit})`);
        return response.reports;
    }
    /**
     * Get reports for multiple feeds at a specific timestamp
     * @param feedIds List of feed IDs to get reports for
     * @param timestamp The timestamp to get reports for
     * @returns Array of raw report data
     * @warning Reports are not guaranteed to be returned in the same order as input feedIds.
     * Always use `report.feedID` to identify each report rather than relying on array position.
     */
    async getReportsBulk(feedIds, timestamp) {
        (0, validation_1.validateFeedIds)(feedIds);
        (0, validation_1.validateTimestamp)(timestamp);
        this.logger.debug(`Fetching bulk reports for ${feedIds.length} feeds at timestamp ${timestamp}`);
        const response = await this.withRetry(() => this.makeRequest(`/api/v1/reports/bulk?feedIDs=${feedIds.join(",")}&timestamp=${timestamp}`));
        this.logger.info(`Retrieved ${response.reports.length} bulk reports for timestamp ${timestamp}`);
        return response.reports;
    }
    /**
     * Create a new Stream instance for real-time data streaming.
     *
     * @param feedIds Feed ID(s) to stream
     * @param options Optional stream configuration
     * @returns Stream instance for real-time report processing
     */
    createStream(feedIds, options) {
        const feedIdArray = Array.isArray(feedIds) ? feedIds : [feedIds];
        feedIdArray.forEach(validation_1.validateFeedId);
        this.logger.debug(`Creating stream for ${feedIdArray.length} feeds: ${feedIdArray.join(", ")}`);
        const streamOptions = options
            ? {
                reconnectInterval: options.reconnectInterval,
                maxReconnectAttempts: options.maxReconnectAttempts,
            }
            : {};
        const stream = new stream_1.Stream(this.config, feedIdArray, streamOptions);
        this.logger.info(`Stream created successfully for ${feedIdArray.length} feed(s)`);
        return stream;
    }
}
exports.DataStreamsClientImpl = DataStreamsClientImpl;
